---
title: "Pipeline_For GeoMX_Data_Analysis"
author: "Thomas Goralski"
date: '2022-07-22'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


This pipeline is intended for expedient analysis of raw data from Nanostring's GeoMx spatial transcriptomics platform.

The following is adapted from the GeoMxTools package Vignettes, and "Spatial Data Analysis Report" from Daniel Newhouse on a data analysis contract. The following includes code/functions adapted from these, my own code, and combinations thereof.

To begin, We load in packages









Now we load our packages
```{r loading packages}
#make list of all packages you will need
```


```{r loading packages}
library_list<- c( "NanoStringNCTools", "GeomxTools", "EnvStats", "ggiraph", "SpatialDecon", "reshape" , "reshape2", "knitr", "dplyr", "ggplot2",  "ggforce","cowplot", "scales", "umap", "Rtsne", "pheatmap", "ggrepel", "rentrez", "RColorBrewer", "Polychrome", "plyr",  "openxlsx", "psych", "kableExtra", "GGally", "magick", "circlize", "pals", "ggupset", "tidyr", "ggpubr", "ggprism", "Biobase", "FactoMineR", "svglite", "corrplot", "pheatmap", "readr", "qusage", "GSEABase", "ggcorrplot","ggbreak","rentrez", "stringr", "writexl", "readxl")

#load in all libraries... Install all packages that error here
get_libraries(library_list)



#If you get an error, there is no package called... Install that package
```



Establish Directories and folders
```{r Create Directories}
# # datadir is the location of dcc files, pkc file(s), and images (if applicable). Ensure these file are in working directory 
 datadir<- getwd()


# Initiate a date tag & directory naming
date_tag <- format(Sys.Date(), "%Y%m%d")

outdir <- paste0("output", date_tag)
object_dir <- file.path(outdir, "R_objects")
qc_dir <- file.path(outdir, "/qc")


# Initialize output directory for static images, tables, and serialized R objects.
dir.create(outdir, recursive = TRUE)  #output directory for results
dir.create(object_dir, recursive = TRUE)  #directory with raw files
dir.create(qc_dir, recursive = TRUE)   #QC output directory
```





we load the data into Nanostrings custom data structure



```{r loading data}
#First check if file exits
if(file.exists(file.path(object_dir, "probe_data.RDS"))){
  probe_data <- readRDS(file.path(object_dir, "probe_data.RDS"))
} else {                                                                #If file does not exist then
 
 DCCFiles <- dir(file.path(datadir, "/DCC_Combined", fsep=""), pattern = ".dcc$",         #specify where the DCC files are (should be datadir) And pattern to  identify 
                 full.names = TRUE, recursive = TRUE)
 PKCFiles <- dir(file.path(datadir, fsep="\\"), pattern = ".pkc$",          #same for PKS file
                                 full.names = TRUE, recursive = TRUE)
 #selct the correct file in paths
 PKCFiles<-PKCFiles[2]
 
 SampleAnnotationFile <-                                                 
  dir(file.path(datadir, fsep="\\" ), pattern = "combined_worksheet_merged_newdata.xlsx",            #specify path and pattern to identify annotation file. ***Make sure it is not opne or else excel creates a temporary save of it, leading to 2 paths
      full.names = TRUE, recursive = TRUE)
 
 SampleAnnotationFile<-SampleAnnotationFile[2]
 
  probe_data <-
    readNanoStringGeoMxSet(dccFiles = DCCFiles,                           #This loads the data in Nanostrings s4 data structure
                           pkcFiles = PKCFiles,
                           phenoDataFile = SampleAnnotationFile,
                           phenoDataSheet = "SegmentProperties",      #full name of annotation file
                                          phenoDataDccColName = "Sample_ID",    #name of column that identifies samples in annotation file
                                          protocolDataColNames = c("aoi", "roi"),  #provide column names in annotation file  aside from panel and sample ID
                                          experimentDataColNames = c("panel"))   #Specify column that identifies panel
  saveRDS(object=probe_data, file=file.path(object_dir, "probe_data.RDS"))
}


#put a column name you know is in your data here, logic check that it all worked
if(!"slide" %in% colnames(pData(probe_data))){
  stop("The reserved column \'slide name\' was not found in this dataset.")
}




```




Next we specify features (genes) of interest, and ensure they are present in the dataset. In addition we identify annotation data that we want to visualize throughout QC (allows to check if specific parameters are causing problems with data), and ensure they are present in the data.

```{r}
#make sure slide is categorical
pData(probe_data)$Slide<-as.character(pData(probe_data)$Slide)

ind1<-which(pData(probe_data)$Slide=="1")

pData(probe_data)$Slide[ind1]<-"A"

ind2<-which(pData(probe_data)$Slide=="2")

pData(probe_data)$Slide[ind2]<-"B"


```

```{r Select most important factors/ffeatures}


foi <- c("Map2", "Rbfox3", "Itgam", "Adgre1", "Aif1", "Snca")   #specify genes ("features") of interest. 

if(any(!foi %in% fData(probe_data)$Target)) {                    #check to see if all foi are in dataset
 foi_not_found <- foi[!foi %in% fData(probe_data)$Target]
 foi_not_found_msg <- paste0("Some features of interest were not found in this dataset. Please check the ",
               "following features to ensure they are correctly entered: ",
               formatList(foi_not_found), ".")
 foi <- foi[foi %in% fData(probe_data)$Target]

}

#enter the factors you'd like to track through QC from annotation data
factors_of_interest <- c("GeoMx_Run", "Mouse","segment", "Layer", "Region") #identify annotation data to track through QC

#define factors of interest that are not of classs numeric for graphing purposes
factors_of_interest_non_numeric<-c("GeoMx_Run" , "segment",  "Region")
#assign factor to color sankey plot by
sankey_focal_factor <- factors_of_interest[1]    #color by segment

gene_detection_rate_color_by<- "GeoMx_Run"


allow_list <- c(foi)


if(any(!factors_of_interest %in% colnames(pData(probe_data)))) {              #check if factors are present in data
 facs_not_found <- factors_of_interest[!factors_of_interest %in% colnames(pData(probe_data))]
 facs_not_found_msg <- paste0("Some factors of interest were not found, please check the ",
               "following features to ensure they are correctly entered: ",
               facs_not_found, ".")
 factors_of_interest <- factors_of_interest[factors_of_interest %in% colnames(pData(probe_data))]
}
# Stop if the focal factor for sankey was not present (e.g., removed in above logic)
if(!sankey_focal_factor %in% factors_of_interest){
  stop("The Sankey focal factor is not present in factors_of_interest.")
}
```


Next we run quality control (QC) for our data. These parameters are highly dependent on the nature of your experiemnt. They can and should be informed by the biological context of your data. For example, in this data we already know that area will be small for pSYN bearing neurons, therefore area is not a very good QC to base removal of data on. I recommend a conservative QC approach to ensure that your analysis is highly likely to be reproducible, but don't needlessly throw data away as this also can bias your analysis.

While exploratory changing QC parameters to ensure you don't selectively remove groups from the data when a biological explanation for QC fail is present is encouraged, DO NOT CHANGE QC PARAMETERS BASED ON SUBSEQUENT DATA ANALYSIS. QC parameters should be determined prior to any data analysis occurance.  

Normalized data will be available in your object directory under the name "target_data"

```{r Set datable Parameters}
#specifc paramters for data tbale outputs from QC analysis
dt_params = 
   list(dom = "lfBtip",
        buttons = list(list(extend = "copy"),
                       list(extend = "csv", filename = "ExampleDataSummary.csv"),
                       list(extend = "excel", filename = "ExampleDataSummary.xlsx")),
        autoWidth = TRUE,
        searching = TRUE,
        scrollX = TRUE,
        pagingType = "simple",
        scrollCollapse = TRUE,
        fixedColumns = list(leftColumns = 1))

```


Setting 
```{r Set QC Parameters}
#specifcy QC parameters
QC_params <- list(
  minSegmentReads = 1000, # segment QC thresholds
  percentTrimmed = 80,
  percentStitched = 80,
  percentAligned = 75,
  percentSaturation = 50,
  minNegativeCount = 1,
  maxNTCCount = 9000,
  minNuclei = 0,   #to recreate dans analysis no nuclei filtering can happen
  minArea = 50,
  minProbeCount = 10, # probe QC thresholds
  minProbeRatio = 0.1,
  outlierTestAlpha = 0.01,
  percentFailGrubbs = 20,
  loqCutoff = 1,
  highCountCutoff = 10000
)


loq_feature_filter_proportion <- 0.01 # feature needs to be in at least this proportion of samples globally
loq_segment_filter_proportion <- 0.03 # Remove samples with low proportion of features above LOQ

#specify column name from pData in probe data file that determines segmentation strategy
index<-which(colnames(pData(probe_data))=="GeoMx_Run")
segment<- colnames(pData(probe_data))[index]
```









Below we get our QC report.The code behind run_qc is highly verbose. I recommend running QC manually until you understand the whole process, then use run_qc as an efficient method for running QC
All plots will be automatically saved to your directory paths in their relevant folder(s). In addition, relevant plots will be printed here to determine any necessary QC tweaks. The function has saved the normalized data to your directory
```{r run QC}


saveRDS(probe_data, file.path(object_dir, "probe_data.RDS"))

 


run_qc(Dataset = probe_data[,], Parameters = QC_params, segment_id = segment)


```









Read in the normalized data and visualize it
```{r Read in Normalized Data}


#reading normalized data
target_data<-readRDS(file.path(object_dir, "target_data.RDS"))






```










Subsetting data...not necessary
```{r}

#first we remove any smaple wth high negative probe detection
ind<-which(rownames(target_data)=="NegProbe-WTX")
 


target_data<-target_data[-ind,]
```



#make compisons of data from runs to ensure similarity
```{r}

A_ind<-which(target_data@phenoData@data$GeoMx_Run=="A")

A<-target_data[,A_ind]


B_ind<-which(target_data@phenoData@data$GeoMx_Run=="B")
B<-target_data[,B_ind]

C_ind<-which(target_data@phenoData@data$GeoMx_Run=="C")
C<-target_data[,C_ind]


#A@phenoData@data$Region%in% B@phenoData@data$Region%in% C@phenoData@data$Region

ind<-A@phenoData@data$Region%in%C@phenoData@data$Region
ind2<-C@phenoData@data$Region%in%A@phenoData@data$Region
A_subC<-A[,ind]
C_subA<-C[,ind2]

summary(C_subA@assayData[["log_q"]])
summary(A_subC@assayData[["log_q"]])

log_q_A<-A_subC@assayData$log_q
colnames(log_q_A)<-A_subC@phenoData@data$GeoMx_Run
  melt_nucq3_A<- melt(log_q_A)
 
log_q_C<-C_subA@assayData$log_q
colnames(log_q_C)<-C_subA@phenoData@data$GeoMx_Run
  melt_nucq3_C<- melt(log_q_C)
 
melt_nucq3<-rbind(melt_nucq3_A,melt_nucq3_C)
 
 p<-ggplot(melt_nucq3, aes(x=Var2, y=value, color=Var2 ))+
  ggtitle("Plot of same regions, q3, log normalized Data")+
  geom_violin(alpha=0.5, bw=0.5)+
    geom_boxplot(width=0.1)
 ggsave(p, filename = file.path(qc_dir, paste0("violin_SlideRun_Compare_AvsC.pdf")), "pdf",
                                width=10, height=6)
  print(p)
  
  
  
  
  
  
  
  
#now do it for b
  ind<-B@phenoData@data$Region%in%C@phenoData@data$Region
ind2<-C@phenoData@data$Region%in%B@phenoData@data$Region
A_subC<-B[,ind]
C_subA<-C[,ind2]

summary(C_subA@assayData[["log_q"]])
summary(A_subC@assayData[["log_q"]])

log_q_A<-A_subC@assayData$log_q
colnames(log_q_A)<-A_subC@phenoData@data$GeoMx_Run
  melt_nucq3_A<- melt(log_q_A)
 
log_q_C<-C_subA@assayData$log_q
colnames(log_q_C)<-C_subA@phenoData@data$GeoMx_Run
  melt_nucq3_C<- melt(log_q_C)
 
melt_nucq3<-rbind(melt_nucq3_A,melt_nucq3_C)
 
 p<-ggplot(melt_nucq3, aes(x=Var2, y=value, color=Var2 ))+
  ggtitle("Plot of same regions, q3, log normalized Data")+
  geom_violin(alpha=0.5, bw=0.5)+
    geom_boxplot(width=0.1)
 ggsave(p, filename = file.path(qc_dir, paste0("violin_SlideRun_Compare_BvsC.pdf")), "pdf",
                                width=10, height=6)
  print(p)

  
  
  summary(C_subA@assayData[["log_q"]][1,])
  
  summary(A_subC@assayData[["log_q"]][1,])
  
  for(i in c(1:100)){
      print(summary(C_subA@assayData[["log_q"]][i,]))
  
  print(summary(A_subC@assayData[["log_q"]][i,]))
  print("NEXT")
  }
  
```





All checks out pretty well

```{r}

pangea_data<-target_data@assayData$log_q

vec<-c(paste(target_data@phenoData@data$Region,"_", target_data@phenoData@data$Layer, "-",target_data@phenoData@data$Mouse,sep="")) 

colnames(pangea_data)<-vec

write.csv(x=pangea_data, file=file.path(object_dir, "Pangea_data2.csv"))

```











Within Slide Analysis: Brain Regions


In this analysis we save log2 fold change estimates and P-values across all levels in the factor of interest. We also apply a Benjamini-Hochberg multiple test correction.

So you can see your data/options for sorting in analysis
```{r}
View(pData(target_data))
```











Begin atlas analysis
```{r}

conam<-paste(target_data@phenoData@data[["Region"]], target_data@phenoData@data[["Layer"]], sep="_")

log_q<-target_data@assayData$log_q

colnames(log_q)<-conam

log_q<-as.data.frame(log_q)

write_csv(log_q, "atlas_data.csv")

```


Now make PCA with ABA colors

```{r}
#read in ABA color scheme
library(readr)
ABA_Colors <- read_csv("ABA_Colors.csv")

#find match to ABA colors in target data
ind<-ABA_Colors$Region %in% target_data@phenoData@data$Region



colors_ABA<-ABA_Colors$Color[ind]
names(colors_ABA)<-ABA_Colors$Region[ind]


names(colors_ABA) %in% target_data@phenoData@data$Region

ind_wrong<-which(target_data@phenoData@data$Region %in%names(colors_ABA)==FALSE) 




names(colors_ABA)
colors_ABA




colors_correction<-c("#FF7AFF","#009FAC", "#08858C", "#FFFC91")
names(colors_correction)<-c("SC", "VISrl/VISal", "VISl/VISli","CUL4,5")

colors_ABA<-c(colors_ABA,colors_correction)




```


```{r}
#Create PCA directory
PCA_plot_dir <-file.path(outdir, "PCA", "Plots")
PCA_data_dir <- file.path(outdir, "PCA", "Data")
dir.create(PCA_plot_dir,recursive = TRUE)
dir.create(PCA_data_dir, recursive = TRUE) 

```



```{r}

#makle dataset with each AOI as rownames, each gene,and pdata as columns
PCA_data<-t(assayDataElement(object = target_data, elt = "log_q"))
PCA_data<- cbind.data.frame(target_data@phenoData@data$Region ,target_data@phenoData@data$Layer, target_data@phenoData@data$AOINucleiCount, target_data@phenoData@data$`slide name`,  target_data@phenoData@data$segment, PCA_data )


#create vector of names of the pData you want in PCA. Order them quantitative, then qualitative
pca_colnames<-c("Region" ,"Layer", "AOINucleiCount" , "slide", "segment")



#place column names
colnames(PCA_data)[1:length(pca_colnames)]<-pca_colnames



quantitative_factors<-c(3)

qualitative_factors<-c(1,2,4,5)


target_PCA<-FactoMineR::PCA(X= PCA_data, 
                ncp=10,                #number of principle components to keep in dataset
                scale.unit = TRUE,   #scales based on z-score... important for PCA, leave true
                ind.sup= NULL,
                quanti.sup= quantitative_factors,    #vector of the indexes of pheno data that is quantitative
                quali.sup = qualitative_factors ,     #vector of indexes of the pheno data that is qualitative
                row.w= NULL,         # weights for rows
                col.w= NULL,         # weights for columns
                graph=FALSE,         #whether graph should be auto displayed
                axes= c(1,2)         # which components to display
                  )



 the_prefix="PCA"

colnames(target_PCA$var$coord) <- gsub("Dim.",
                                  paste0(the_prefix, "var_coords_"), colnames(target_PCA$ind$coord)) 
 
  
ind_pca<-target_PCA[["ind"]][["coord"]]



colors<-c("blue","green","sandybrown","orange","brown","gray","purple","pink","black","cyan","magenta","violet")



region_match<-tolower(target_data@phenoData@data$Region)

library(readxl)
Special_Region_Designation <- read_excel("Special Region Designation.xlsx")

ind_false<-which(region_match %in% Special_Region_Designation$Region ==FALSE)

newregion_colors <- readRDS("~/Henderson_Lab/Geo_Atlas_Combo_Pangea/GeoAtlas_Merge/attempt3.rds")
ColorMatch <- readRDS("~/Henderson_Lab/Geo_Atlas_Combo_Pangea/GeoAtlas_Merge/ColorMatch.rds")

target_data@phenoData@data$newRegion<-newregion_colors$major

color_names<-ColorMatch$X11
names(color_names)<-ColorMatch$X1

p <- ggplot(data=target_data@phenoData@data, 
            aes(x=ind_pca[,1], y=ind_pca[,2])) +
            geom_point(aes(color=target_data@phenoData@data$newRegion), alpha=0.8, size=0.5, stroke=0.2) + 
            labs(x=paste0("PCA 1 (", round(target_PCA$eig[1,2]), "%)"), 
            y=paste0("PCA 2 (", round(target_PCA$eig[2,2]), "%)"),
                        title="PCA") +
            theme_bw(base_size=2) +
            theme(legend.position = "right",  
                  legend.text = element_text( size = 0.5),
                  plot.background = element_blank(),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank(),
                  legend.key.size = unit(0.15, 'cm')
                  )+
  scale_color_manual(values=color_names)+
  guides(shape = guide_legend(override.aes = list(size = 0.01)))+
  guides(color = guide_legend(override.aes = list(size = 0.01)))+
  theme(legend.title = element_text(size = 0.5), 
               legend.text = element_text(size = 0.5))
 
p
   
regions<-target_data@phenoData@data$Region
saveRDS(regions,  "Regions.RDS")

 ggsave("PCA_newcolors.pdf", p,  width = 10, height = 6, units = "cm",path = PCA_plot_dir)

```



Now make PCA with only cortical regions




```{r}

target_data_cortex<-target_data[,which(target_data@phenoData@data$Region %in% c("ACA", "ACAd", "ACAv", "SSp", "SSs", "SSp-bfd", "SSp-ll", "SSp-m", "SSp-n", "SSp-tr", "SSp-ul", "SSs-bfd", "SSs-ll", "SSs-m", "SSs-n", "SSs-tr", "SSs-ul","ORBl",  "ORBvl" ,"ORBm", "PL","MOs","FRP","AId","AIv", "MOp", "SSp", "VISC", "GU","AIp", "CLA","EPd", "EPv", "PIR", "RSPv","RSPd", "RSPagl", "SSs", "AUDv","TEa","NLOT",  "COAa","PAA", "AUDp","ECT", "PERI",  "LA",  "BLAp", "BLAa", "BLAv", "BMAp", "PA", "COApl", "COApm", "CA1",  "CA3", "DG", "POST", "PRE", "PAR", "ENTm", "ENTl", "VISp" , "SUB", "ProS", "HATA", "BMA"  , "VISpm", "VISrl/VISal", "DG-po", "DG-sg", "TR", "VISl/VISli","MOB", "TTd","TTv","AON", "AOB","VISa" ))]




#makle dataset with each AOI as rownames, each gene,and pdata as columns
PCA_data_cortex<-t(assayDataElement(object = target_data_cortex, elt = "log_q"))
PCA_data_cortex<- cbind.data.frame(target_data_cortex@phenoData@data$Region ,target_data_cortex@phenoData@data$Layer, target_data_cortex@phenoData@data$AOINucleiCount, target_data_cortex@phenoData@data$`slide name`,  target_data_cortex@phenoData@data$segment, PCA_data_cortex )


#create vector of names of the pData you want in PCA. Order them quantitative, then qualitative
pca_colnames<-c("Region" ,"Layer", "AOINucleiCount" , "slide", "segment")



#place column names
colnames(PCA_data_cortex)[1:length(pca_colnames)]<-pca_colnames



quantitative_factors<-c(3)

qualitative_factors<-c(1,2,4,5)


target_PCA<-FactoMineR::PCA(X= PCA_data_cortex, 
                ncp=10,                #number of principle components to keep in dataset
                scale.unit = TRUE,   #scales based on z-score... important for PCA, leave true
                ind.sup= NULL,
                quanti.sup= quantitative_factors,    #vector of the indexes of pheno data that is quantitative
                quali.sup = qualitative_factors ,     #vector of indexes of the pheno data that is qualitative
                row.w= NULL,         # weights for rows
                col.w= NULL,         # weights for columns
                graph=FALSE,         #whether graph should be auto displayed
                axes= c(1,2)         # which components to display
                  )



 the_prefix="PCA"

colnames(target_PCA$var$coord) <- gsub("Dim.",
                                  paste0(the_prefix, "var_coords_"), colnames(target_PCA$ind$coord)) 
 
  
ind_pca<-target_PCA[["ind"]][["coord"]]



colors<-c("blue","green","sandybrown","orange","brown","gray","purple","pink","black","cyan","magenta","violet")





p <- ggplot(data=target_data_cortex@phenoData@data, 
            aes(x=ind_pca[,1], y=ind_pca[,2])) +
            geom_point(aes(color=target_data_cortex@phenoData@data$Region), alpha=0.5, size=0.5, stroke=0.2) + 
            labs(x=paste0("PCA 1 (", round(target_PCA$eig[1,2]), "%)"), 
            y=paste0("PCA 2 (", round(target_PCA$eig[2,2]), "%)"),
                        title="PCA") +
            theme_bw(base_size=2) +
            theme(legend.position = "right",  
                  legend.text = element_text( size = 0.5),
                  plot.background = element_blank(),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank(),
                  legend.key.size = unit(0.15, 'cm')
                  )+
  scale_color_manual(values=colors_ABA)+
  guides(shape = guide_legend(override.aes = list(size = 0.01)))+
  guides(color = guide_legend(override.aes = list(size = 0.01)))+
  theme(legend.title = element_text(size = 0.5), 
               legend.text = element_text(size = 0.5))
 
p


ggsave("PCA_Cortex_Only.pdf", p,  width = 10, height = 6, units = "cm",path = PCA_plot_dir)

```



Now do non-cortical regions
```{r}

`%notin%` <- function (x, table) is.na(match(x, table, nomatch=NA_integer_))

target_data_noncortex<-target_data[,which(target_data@phenoData@data$Region %notin% c("ACA", "ACAd", "ACAv", "SSp", "SSs", "SSp-bfd", "SSp-ll", "SSp-m", "SSp-n", "SSp-tr", "SSp-ul", "SSs-bfd", "SSs-ll", "SSs-m", "SSs-n", "SSs-tr", "SSs-ul","ORBl",  "ORBvl" ,"ORBm", "PL","MOs","FRP","AId","AIv", "MOp", "SSp", "VISC", "GU","AIp", "CLA","EPd", "EPv", "PIR", "RSPv","RSPd", "RSPagl", "SSs", "AUDv","TEa","NLOT",  "COAa","PAA", "AUDp","ECT", "PERI",  "LA",  "BLAp", "BLAa", "BLAv", "BMAp", "PA", "COApl", "COApm", "CA1",  "CA3", "DG", "POST", "PRE", "PAR", "ENTm", "ENTl", "VISp" , "SUB", "ProS", "HATA", "BMA"  , "VISpm", "VISrl/VISal", "DG-po", "DG-sg", "TR", "VISl/VISli","MOB", "TTd", "TTv", "AON", "AOB","VISa"))]




#makle dataset with each AOI as rownames, each gene,and pdata as columns
PCA_data_noncortex<-t(assayDataElement(object = target_data_noncortex, elt = "log_q"))
PCA_data_noncortex<- cbind.data.frame(target_data_noncortex@phenoData@data$Region ,target_data_noncortex@phenoData@data$Layer, target_data_noncortex@phenoData@data$AOINucleiCount, target_data_noncortex@phenoData@data$`slide name`,  target_data_noncortex@phenoData@data$segment, PCA_data_noncortex )


#create vector of names of the pData you want in PCA. Order them quantitative, then qualitative
pca_colnames<-c("Region" ,"Layer", "AOINucleiCount" , "slide", "segment")



#place column names
colnames(PCA_data_noncortex)[1:length(pca_colnames)]<-pca_colnames



quantitative_factors<-c(3)

qualitative_factors<-c(1,2,4,5)


target_PCA<-FactoMineR::PCA(X= PCA_data_noncortex, 
                ncp=10,                #number of principle components to keep in dataset
                scale.unit = TRUE,   #scales based on z-score... important for PCA, leave true
                ind.sup= NULL,
                quanti.sup= quantitative_factors,    #vector of the indexes of pheno data that is quantitative
                quali.sup = qualitative_factors ,     #vector of indexes of the pheno data that is qualitative
                row.w= NULL,         # weights for rows
                col.w= NULL,         # weights for columns
                graph=FALSE,         #whether graph should be auto displayed
                axes= c(1,2)         # which components to display
                  )



 the_prefix="PCA"

colnames(target_PCA$var$coord) <- gsub("Dim.",
                                  paste0(the_prefix, "var_coords_"), colnames(target_PCA$ind$coord)) 
 
  
ind_pca<-target_PCA[["ind"]][["coord"]]



colors<-c("blue","green","sandybrown","orange","brown","gray","purple","pink","black","cyan","magenta","violet")


p <- ggplot(data=target_data_noncortex@phenoData@data, 
            aes(x=ind_pca[,1], y=ind_pca[,2])) +
            geom_point(aes(color=target_data_noncortex@phenoData@data$Region), alpha=0.5, size=0.5, stroke=0.2) + 
            labs(x=paste0("PCA 1 (", round(target_PCA$eig[1,2]), "%)"), 
            y=paste0("PCA 2 (", round(target_PCA$eig[2,2]), "%)"),
                        title="PCA") +
            theme_bw(base_size=2) +
            theme(legend.position = "right",  
                  legend.text = element_text( size = 0.5),
                  plot.background = element_blank(),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank(),
                  legend.key.size = unit(0.15, 'cm')
                  )+
  scale_color_manual(values=colors_ABA)+
  guides(shape = guide_legend(override.aes = list(size = 0.01)))+
  guides(color = guide_legend(override.aes = list(size = 0.01)))+
  theme(legend.title = element_text(size = 0.5), 
               legend.text = element_text(size = 0.5))
 
p


ggsave("PCA_Non_Cortex.pdf", p,  width = 10, height = 6, units = "cm",path = PCA_plot_dir)






```




Now get a heatmap of each region for the vulnerability associated genes (1616 genes in correlations). 

```{r}




```




Do correlation with allen in situ atlas

```{r}
library(ggplot2)
library(reshape2)
library(data.table)
library(ggpubr)
library(readr)
#read in allen data

AllGeneExpression_Fulcher2019 <- read_csv("AllGeneExpression_Fulcher2019.csv")

AllGeneExpression_Fulcher2019<-as.data.frame(AllGeneExpression_Fulcher2019)
rownames(AllGeneExpression_Fulcher2019)<-AllGeneExpression_Fulcher2019$...1
AllGeneExpression_Fulcher2019$region<-AllGeneExpression_Fulcher2019$...1

#get matching layout dataset
Pangea<-target_data@assayData[["log_q"]]

colnames(Pangea)<-target_data@phenoData@data$Region
Pangea<-t(Pangea)




pangea_allen_overlap<-merge(Pangea, AllGeneExpression_Fulcher2019, by="row.names")

conserved_regions<-pangea_allen_overlap$Row.names


Pangea<-Pangea[conserved_regions,]

AllGeneExpression_Fulcher2019<-AllGeneExpression_Fulcher2019[conserved_regions,]



ind_col<-colnames(Pangea) %in% colnames(AllGeneExpression_Fulcher2019)

Pangea<-Pangea[,ind_col]

ind_col<-colnames(AllGeneExpression_Fulcher2019) %in% colnames(Pangea)

AllGeneExpression_Fulcher2019<-AllGeneExpression_Fulcher2019[,ind_col]


AllGeneExpression_Fulcher2019$region<-rownames(AllGeneExpression_Fulcher2019)

allen_melt<-melt(AllGeneExpression_Fulcher2019)

Pangea<-as.data.frame(Pangea)
Pangea$region<-rownames(Pangea)
pangea_melt<-melt(Pangea)


allen_melt$dataset<-"allen"
pangea_melt$dataset<-"pangea"



merge_allen_pangea<-merge(allen_melt, pangea_melt, by=c("region","variable"))



results_lm<-lm(merge_allen_pangea$value.x ~ merge_allen_pangea$value.y, data=merge_allen_pangea)
summary(results_lm)






#find match to ABA colors in target data
ind<-ABA_Colors$Region %in% merge_allen_pangea$region
colors_ABA<-ABA_Colors$Color[ind]
names(colors_ABA)<-ABA_Colors$Region[ind]
names(colors_ABA) %in% merge_allen_pangea$region
ind_wrong<-which(merge_allen_pangea %in%names(colors_ABA)==FALSE) 
names(colors_ABA)
colors_ABA
colors_correction<-c("#FF7AFF","#009FAC", "#08858C", "#FFFC91")
names(colors_correction)<-c("SC", "VISrl/VISal", "VISl/VISli","CUL4,5")
colors_ABA<-c(colors_ABA,colors_correction)
melt_colors_aba<-as.data.frame(colors_ABA)
melt_colors_aba$region<-rownames(melt_colors_aba)

merge_allen_pangea<-merge(merge_allen_pangea, melt_colors_aba, by="region")

merge_allen_pangea$colors<-merge_allen_pangea$colors_ABA


y_axis_eq<-seq(from=13, to=30, by=0.05)
 
x_axis_eq <-c(0,0.2,0.4,0.6,0.8)

x_axis_eq<-rep(x_axis_eq, length(y_axis_eq))

p <- ggplot(merge_allen_pangea[c(1:200),],aes(x=value.x, y= value.y,color=merge_allen_pangea$colors)) + 
    geom_point(alpha = 0.25, size = 1,  ,shape=16)+ #color = merge_allen_pangea$region)+ 
    labs(x ="Log2FC", y = "Log2FC", title = "PANGEA correlation to Allen In Situ Atlas") + 
    geom_smooth(method = "lm", formula = y ~ x,se=FALSE) +
   stat_regline_equation(aes(label = paste(..eq.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, label.x = x_axis_eq, label.y = y_axis_eq, size=1)+
  theme(legend.position = "none")+
  facet_wrap(~variable, scales = "free")
  #scale_color_manual(values=colors_ABA)

 
p

ggsave("Correlation_Pangea_AllenInSitu_By_Gene.pdf",plot=p, width = 1040, height = 1040, units = "cm", limitsize = FALSE)





p <- ggplot(merge_allen_pangea[c(1:200),],aes(x=value.x, y= value.y,color=merge_allen_pangea$colors)) + 
    geom_point(alpha = 0.25, size = 1,  ,shape=16)+ #color = merge_allen_pangea$region)+ 
    labs(x ="Log2FC", y = "Log2FC", title = "PANGEA correlation to Allen In Situ Atlas") + 
    geom_smooth(method = "lm", formula = y ~ x,se=FALSE) +
   stat_regline_equation(aes(label = paste(..eq.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, label.x = x_axis_eq, label.y = y_axis_eq, size=1)+
  theme(legend.position = "none")+
  facet_wrap(~variable, scales = "free")
  #scale_color_manual(values=colors_ABA)

 
p

```




Now we do it with the sigmoid tranformed PANGEA data. 
```{r}

norm_avg_Pangea_exp <- read_csv("norm_avg_Pangea_exp.csv")

norm_avg_Pangea_exp<-as.data.frame(norm_avg_Pangea_exp)
rownames(norm_avg_Pangea_exp)<-norm_avg_Pangea_exp$...1


norm_avg_Pangea_exp$region<-norm_avg_Pangea_exp$...1
norm_avg_Pangea_exp_melt<-melt(norm_avg_Pangea_exp)
norm_avg_Pangea_exp_melt$dataset<-"pangea"


merge_allen_pangea_sigmoid<-merge(allen_melt, norm_avg_Pangea_exp_melt, by=c("region","variable"))


merge_allen_pangea_sigmoid<-merge(merge_allen_pangea_sigmoid, melt_colors_aba, by="region")

merge_allen_pangea_sigmoid$colors<-merge_allen_pangea_sigmoid$colors_ABA

#no facet
p <- ggplot(merge_allen_pangea_sigmoid[,],aes(x=value.x, y= value.y,color=merge_allen_pangea_sigmoid$colors_ABA.x[])) + 
    geom_point(alpha = 0.5, size = 1,  ,shape=16, color=merge_allen_pangea_sigmoid$colors_ABA)+ #color = merge_allen_pangea$region)+ 
    labs(x ="Log2FC", y = "Log2FC", title = "PANGEA correlation to Allen In Situ Atlas") + 
    geom_smooth(method = "lm", formula = y ~ x,se=FALSE) +
  stat_regline_equation(aes(label = paste(..eq.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, label.x = 0.25, label.y = 1.05, size=6)+
  theme(legend.position = "none")
 # facet_wrap(~variable, scales = "free")
  #scale_color_manual(values=colors_ABA)

 
p

ggsave("Correlation_Pangea_AllenInSitu_All_Gene_sigmoid.pdf",plot=p, width = 20, height = 15, units = "cm", limitsize = FALSE)



p <- ggplot(merge_allen_pangea_sigmoid[,],aes(x=value.x, y= value.y,color=merge_allen_pangea_sigmoid$colors_ABA.x[])) + 
    geom_point(alpha = 0.5, size = 1,  ,shape=16)+ #color = merge_allen_pangea$region)+ 
    labs(x ="Log2FC", y = "Log2FC", title = "PANGEA correlation to Allen In Situ Atlas") + 
    geom_smooth(method = "lm", formula = y ~ x,se=FALSE) +
  # stat_regline_equation(aes(label = paste(..eq.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, label.x = 0.25, label.y = 1.2, size=1)+
  theme(legend.position = "none")+
  facet_wrap(~variable, scales = "free")
  #scale_color_manual(values=colors_ABA)

 
p

ggsave("Correlation_Pangea_AllenInSitu_By_Gene_sigmoid.pdf",plot=p, width = 1040, height = 1040, units = "cm", limitsize = FALSE)



goi<-c("Lrrk2", "Cux2","Gad1","Snca", "Rprm", "Etv1", "Slc17a7","Gad2", "Th")


goi_ind<-which(merge_allen_pangea_sigmoid$variable%in% goi)

p <- ggplot(merge_allen_pangea_sigmoid[goi_ind, ],aes(x=value.x, y= value.y)) + 
    geom_point(alpha = 1, size = 2,  ,shape=16, color=merge_allen_pangea_sigmoid$colors_ABA[goi_ind])+ #color = merge_allen_pangea$region)+ 
    labs(x ="Log2FC", y = "Log2FC", title = "PANGEA correlation to Allen In Situ Atlas") + 
    geom_smooth(method = "lm", formula = y ~ x,se=FALSE) +
   stat_regline_equation(aes(label = paste( ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, label.x = 0.25, label.y = 1, size=4)+
  theme(legend.position = "none")+
  facet_wrap(~variable, scales = "free")+
  theme_bw()+
  theme(legend.position = "right",  
                  legend.text = element_text( size = 0.5),
                  plot.background = element_blank(),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank()
        )
  #scale_color_manual(values=colors_ABA)

 
p

ggsave("Correlation_Pangea_AllenInSitu_By_Gene_sigmoid_goi.pdf",plot=p, width = 30, height = 20, units = "cm", limitsize = FALSE)




# p <- ggplot(merge_allen_pangea_sigmoid[, ],aes(x=value.x, y= value.y,color=merge_allen_pangea_sigmoid$colors[])) + 
#     geom_point(alpha = 0.25, size = 1,  ,shape=16)+ #color = merge_allen_pangea$region)+ 
#     labs(x ="Log2FC", y = "Log2FC", title = "PANGEA correlation to Allen In Situ Atlas") + 
#     geom_smooth(method = "lm", formula = y ~ x,se=FALSE) +
#    stat_regline_equation(aes(label = paste(..eq.label.., ..adj.rr.label.., sep = "~~~~")), formula = y ~ x, label.x = x_axis_eq, label.y = y_axis_eq, size=1)+
#   theme(legend.position = "none")
#   #scale_color_manual(values=colors_ABA)
# 
#  
# p
# 
# ggsave("Correlation_Pangea_AllenInSitu_sigmoid.pdf",plot=p, width = 60, height = 60, units = "cm", limitsize = FALSE)






```



Now get corr plot for pangea vs aba
```{r}
# norm_avg_Pangea_exp
# 
# norm_avg_Pangea_exp<-as.data.frame(norm_avg_Pangea_exp)
# rownames(norm_avg_Pangea_exp)<-norm_avg_Pangea_exp$...1
# 
# 
# norm_avg_Pangea_exp$region<-norm_avg_Pangea_exp$...1
# norm_avg_Pangea_exp_melt<-melt(norm_avg_Pangea_exp)
# norm_avg_Pangea_exp_melt$dataset<-"pangea"
# 
# 
# library(ggcorrplot)
# 
# norm_avg_pangea_corrdat<-norm_avg_Pangea_exp[,-1]
# ind_cor<-colnames(norm_avg_pangea_corrdat) %in%colnames(AllGeneExpression_Fulcher2019)
# norm_avg_pangea_corrdat<-norm_avg_pangea_corrdat[,ind_cor]
# ind_cor2<-colnames(AllGeneExpression_Fulcher2019) %in% colnames(norm_avg_pangea_corrdat)
# AllGeneExpression_Fulcher2019_overlap<-AllGeneExpression_Fulcher2019[,ind_cor2]
# 
# ind_cor3<-rownames(AllGeneExpression_Fulcher2019_overlap) %in% rownames(norm_avg_pangea_corrdat)
# 
# AllGeneExpression_Fulcher2019_overlap<-AllGeneExpression_Fulcher2019_overlap[ind_cor3,]
# ind_cor4<- rownames(norm_avg_pangea_corrdat) %in% rownames(AllGeneExpression_Fulcher2019_overlap)
# 
# norm_avg_pangea_corrdat<-norm_avg_pangea_corrdat[ind_cor4,]
# 
# 
# 
# pan_allen_cor<-cor(norm_avg_pangea_corrdat, AllGeneExpression_Fulcher2019_overlap)
# 
# p<-ggcorrplot(pan_allen_cor[c(3665:3665),c(3665:3665)],type = "lower", outline.color = NA,show.diag = TRUE, tl.cex = 0.4, hc.order = TRUE)+
#   theme(axis.line = element_line(color='black'),
#     plot.background = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     panel.border = element_blank())
# 
# -3664,-3664
# 
# 
# 
# 
# p<-ggcorrplot(pan_allen_cor[-c(3664:3880),-c(3664:3880)],type = "lower", outline.color = NA,show.diag = TRUE, tl.cex = 0.4, hc.order = TRUE)+
#   theme(axis.line = element_line(color='black'),
#     plot.background = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     panel.border = element_blank())
# 
# 
# ggsave("Correlation_plot_pangea_allen.pdf",plot=p, width = 40, height = 40, units = "cm", limitsize = FALSE)
# 
# 
# diag_dat<-c()
# for(i in rownames(pan_allen_cor)){
#   print(i)
#   ind<-which(colnames(pan_allen_cor)==i)
#   diag_ind<-pan_allen_cor[i,ind]
#   names(diag_ind)<-i
# diag_dat<-c(diag_dat, diag_ind)
# 
# }
# 
# 
# 
# #pan_allen_cor_diag<-pan_allen_cor[,ind_diag]
# 
# vec<-rep(0,3826)
# diag_dat<-rbind(diag_dat,vec)
# 
# diag_dat_trans<-t(diag_dat)
# diag_dat_trans<-as.data.frame(diag_dat_trans)
# 
# ind_diag<-order(diag_dat_trans$diag_dat)
# 
# 
# pheatmap(diag_dat_trans[ind_diag,], fontsize_row = 0.5, cluster_rows = FALSE)
# 
# diag_dat<-as.data.frame(diag_dat)
# 
# melt_diag<-melt(diag_dat)
# 
# ind_neg<-which(melt_diag$value<0)
# 
# melt_diag_neg<-melt_diag[ind_neg,]
# 
# 
# 
# 
# ind_aba<-colnames(AllGeneExpression_Fulcher2019_overlap) %in%melt_diag_neg$variable
# 
# AllGeneExpression_Fulcher2019_overlap_negcor<-AllGeneExpression_Fulcher2019_overlap[,ind_aba]
# 
# ind_pangea<-colnames(norm_avg_pangea_corrdat) %in% melt_diag_neg$variable
# 
# norm_avg_pangea_corrdat_negcor<-norm_avg_pangea_corrdat[ind_pangea,]
# 
# length(which(diag_dat_trans$diag_dat>0))
# 
# 
# #83.2% of gene are positively correlated. Interestingly that means that 16.8% of genes are negatively correlated or not correlated at all.
# 3183/length(diag_dat_trans$diag_dat)
# 
# 
# #median correlation is 0.25474,mean is 0.27039. Min is -0.50061. Max is 0.91811. 
# summary(diag_dat_trans$diag_dat)
# 
# 
# 
# 
# breaks<-seq(-1,1,0.005)
# 
# length(breaks)



library(ggforce)
#remove regions column
norm_avg_pangea_corrdat<-norm_avg_pangea_corrdat[,-3827]
AllGeneExpression_Fulcher2019_overlap<-AllGeneExpression_Fulcher2019_overlap[,-3827]

#try again
diag_dat<-c()
for(i in 1:length(norm_avg_pangea_corrdat)){
genename<-colnames(norm_avg_pangea_corrdat)[i]
ind_fulcher<-which(colnames(AllGeneExpression_Fulcher2019_overlap)==genename)
pan_dat<-norm_avg_pangea_corrdat[,i]
fulcher_dat<-AllGeneExpression_Fulcher2019_overlap[,ind_fulcher]
cor_value<-cor(pan_dat, fulcher_dat)
dat_vec<-c(genename, cor_value)
diag_dat<-rbind(diag_dat, dat_vec)
}

diag_dat2<-diag_dat
diag_dat2<- diag_dat2 %>% na.omit()
diag_dat2<-as.data.frame(diag_dat2)
diag_pivot<-t(diag_dat2)
colnames(diag_pivot)<-diag_dat2$V1


diag_dat2$color<-c()



p<-ggplot(data=melt_diag2 ) +
  geom_jitter(aes(y=melt_diag2$value, x=0, color=melt_diag2$color),alpha=0.3, stroke=NA, height =0)+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=0.5))+
  scale_color_manual(values=c(below="blue",above= "green3"))+
  theme_bw()

p


diag_dat2$color<-"blue"
diag_dat2$color[which(diag_dat2$V2>0)]<-"green3"
diag_dat2$group<-"A"

diag_dat2$V2<-as.numeric(diag_dat2$V2)
diag_dat2$correlation<-diag_dat2$V2

p<-ggplot(data=diag_dat2, aes(y=diag_dat2$correlation, x=diag_dat2$group, )) +
  geom_sina(alpha=0.3, stroke=NA, colour=diag_dat2$color  )+
   geom_violin(alpha=0.3)+
  theme(axis.text.y = element_blank())+
  theme_bw()+
  ylim(-1,1)


p


ggsave("violin_jitter.pdf",plot=p, width = 10, height = 15, units = "cm", limitsize = FALSE)




# p<-pheatmap(diag_pivot, fontsize_row = 0.5, cluster_rows = FALSE, color=colorRampPalette(c("blue", "white", "red"))(401),breaks=breaks)
# 
# 
# ggsave("PANGEA_allen_corr_heatmap.pdf",plot=p, width = 10, height = 15, units = "cm", limitsize = FALSE)
# 
# 
# 
# summary(diag_dat_trans[ind_diag,])
# 
# 
# 
# barplot(melt_diag$value)
# 
# melt_diag2<-melt_diag
# melt_diag2<-melt_diag2 %>% na.omit()
# 
# 
# melt_diag2$color<-c()
# 
# ind_below<-which(melt_diag2$value<0)
# ind_above<-which(melt_diag2$value>0)
# melt_diag2$color<-"below"
# melt_diag2$color[ind_above]<-"above"
# 
# p<-ggplot(data=melt_diag2, aes(x=variable, y=value)) +
#   geom_bar(stat="identity",color=melt_diag2$color)+
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=0.5) )
# p
# 
# 
# 
# p<-ggplot(data=melt_diag2 ) +
#   geom_jitter(aes(y=melt_diag2$value, x=0, color=melt_diag2$color),alpha=0.3, stroke=NA, height =0)+
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=0.5))+
#   scale_color_manual(values=c(below="blue",above= "green3"))+
#   theme_bw()
# 
# p
# 
# 
# 
# 
# p<-ggplot(data=melt_diag2, aes(y=melt_diag2$value, x=0) ) +
#   geom_violin()+
#    geom_point(aes(y=melt_diag2$value, x=0, color=melt_diag2$color),alpha=0.3, stroke=NA)+
#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=0.5))+
#   scale_color_manual(values=c(below="blue",above= "green3"))+
#   theme_bw()
# 
# p


write.csv(melt_diag, "melt_diag.csv")



```


Find out what the negatively correlated genes are 

```{r}

ind_neg<-which(diag_dat2$color=="blue")

neg_corr<-diag_dat2[ind_neg,]

write_csv(neg_corr,"neg_corr_genes.csv")

```
















































Convert ACAv and ACAd Regions to just ACA


```{r}


#Convert ACAv and ACAd Regions to just ACA

regions <- target_data@phenoData@data[["region"]]

ind_ACA<-grep(".*ACA.*", regions)

pData(target_data)$region[ind_ACA]="ACA"





```








Get rid of layer two, because of small number of samples
```{r}
#only run once
ind_L2<-which(target_data@phenoData@data$layer==2)

target_data<-target_data[,-ind_L2]
```
















Get heatmap of all genes
```{r}


annots<-target_data@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 


max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)

#color_pal<-colorRampPalette(c("black", "purple2", "yellow2"))(121)

color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(80))


#color_pal<-(colorRampPalette(c( "white","steelblue1", "firebrick1"))(121))


#color_pal<-(colorRampPalette(c("#0092b5", "red"))(301))



heat_dat_maxmin<-target_data@assayData$log_q



#heat_dat_maxmin[which(heat_dat_maxmin>7)]<-7

#heat_dat_maxmin[which(heat_dat_maxmin<1)]<-1



p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = FALSE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal
)


p



genes_plot_dir <-file.path(outdir, "Genes", "Plots")
genes_data_dir <- file.path(outdir, "Genes", "Data")
dir.create(genes_plot_dir,recursive = TRUE)
dir.create(genes_data_dir, recursive = TRUE)



ggsave("GenesHeatmap.png",p, width = 14, height = 14, units = "cm",path = genes_plot_dir)
ggsave("GenesHeatmap.pdf",p, width = 14, height = 14, units = "cm",path = genes_plot_dir)




```













Define Function for formatting LMM results

```{r}
formatLMMResults <- function(lmm_results, p_adjust_method="fdr") {
 if(!inherits(lmm_results, "matrix")){
  stop("lmm_results needs be a matrix, See ?GeomxTools::mixedModelDE.")
 }
 if(!all(c("anova", "lsmeans") %in% rownames(lmm_results))){
  stop("Expected row names of lmm_results to have anova and lsmeans.")
 }
 df <- do.call(rbind, lmm_results["lsmeans", ])
 contrasts <- rownames(df)

 # Make sure there are not multiple " - " present in contrasts
 for(i in unique(contrasts)){
   if(length(strsplit(i, split=" - ")[[1]])>2){
     stop(paste0("Contrast \'", i, "\' has more than two split points. Please rename contrasts first."))
   }
 }
 # Contrast are of the for B - A. Convert to comparisons of the form
 # A vs B.
 df <- as.data.frame(df)
 contrast_pairs <- strsplit(contrasts, split=" - ")
 df$Comparison <- paste0(unlist(lapply(contrast_pairs, "[[", 2L)), " vs ", unlist(lapply(contrast_pairs, "[[", 1L)))
 colnames(df)[which(names(df) == "Pr(>|t|)")] <- "P"
 row.names(df) <- NULL

 # Add feature names
 df$Feature <- rep(colnames(lmm_results), each=nrow(lmm_results["lsmeans",][[1]]))

 # P-adjustment based on subsets of data faceted by Comparison
 df <- ddply(df, .(Comparison), function(x){
   x$padj <- p.adjust(x$P, method = p_adjust_method)
   return(x)
 })
 df <- df[, c("Feature", "Comparison", "Estimate","P", "padj")]
 colnames(df)[colnames(df)=="padj"] <- toupper(p_adjust_method) # 'FDR' used in standard Report


 return(df)
}
```












#Took about 1.5 hours to run
```{r}

# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$segment, c("pSyn", "NeuN"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))

vec<-unique(target_data@phenoData@data$layer)
pData(target_data)[["Layer"]] <- 
    factor(pData(target_data)$layer, c(vec))
# run LMM:
# formula follows conventions defined by the lme4 package
results <- c()
for(status in c( "ACA", "MOs" , "MOp")) {
    ind <- which(pData(target_data)$region == status)
    mixedOutmc <-
        mixedModelDE(target_data[, ind],
                     elt = "log_q",
                     modelFormula = ~ testRegion * Layer + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results1 <-formatLMMResults(mixedOutmc)
    results1$contrast<-status
    
    results <- rbind(results,results1)
    print(status)
}



results$Gene<-results$Feature


#remove neg probe from data
sub<-grep(".*NegP.*", results$Feature)
results<-results[-sub,]


#results$Color<-c()

# Categorize Results based on P-value & FDR for plotting
#results$Color[results$FDR > 0.05  ] <- "NS or FC < 0.5"

results$Color[results$P < 0.05 & results$Estimate>0] <- "Enriched in pSyn P < 0.05"
results$Color[results$P < 0.05 & results$Estimate<0] <- "Enriched in NeuN P < 0.05"

results$Color[results$FDR < 0.05 & results$Estimate>0] <- "Enriched in pSyn FDR < 0.05"
results$Color[results$FDR < 0.05 & results$Estimate<0] <- "Enriched in NeuN FDR < 0.05"


results$Color[results$FDR < 0.01 & results$Estimate>0 ] <- "Enriched in pSyn FDR < 0.01"
results$Color[results$FDR < 0.01 & results$Estimate<0 ] <- "Enriched in NeuN FDR < 0.01"

results$Color[abs(results$Estimate) < 0.5] <- "NS or FC < 0.5"


results$Color <- factor(results$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in pSyn FDR < 0.05", "Enriched in NeuN FDR < 0.05","Enriched in pSyn P < 0.05", "Enriched in NeuN P < 0.05", "Enriched in pSyn FDR < 0.01", "Enriched in NeuN FDR < 0.01" ))

# pick top genes for either side of volcano to label
# order genes for convenience:
results$invert_P <- (-log10(results$P)) * sign(results$Estimate)

top_g <- c()

   for(i in unique(results$contrast)){
   vec<-which(results$contrast==i) 
top_g <- c(top_g,
  results[vec, 'Gene'][order(results[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results[vec, 'Gene'][order(results[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)

highlight_top_g<-subset(results, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

# Graph results
diff_exp3<-ggplot(results,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)


de_plot_dir <-file.path(outdir, "DE", "Plots")
de_data_dir <- file.path(outdir, "DE", "Data")
dir.create(de_plot_dir,recursive = TRUE)
dir.create(de_data_dir, recursive = TRUE)


saveRDS(diff_exp3, file=file.path(de_data_dir, "DE_by_region.RDS"))
ggsave("volc_region.png", width = 14, height = 7, units = "cm",path = de_plot_dir)
ggsave("volc_region.pdf", width = 14, height = 7, units = "cm",path = de_plot_dir)


diff_exp3


```




Violins
```{r}
#begin violin plotting
prop<-target_data@assayData$log_q
annots<-target_data@phenoData@data

violin_df <- cbind(annots %>% dplyr::select(eval(segment), mouse, region),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2,3), names_to = "Feature", values_to = "Expression")


violin_df$contrast<-violin_df$region


violin_p_df <- filter(results, Feature %in% top_g)  
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2

violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

violin_df<-base::merge(violin_p_df, violin_df, by= c("Feature","contrast"))

violin_df$FDR<-signif(violin_df$FDR, 3)

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, color = NA) +
  geom_jitter(width=0.1, height=0, size = 0.5, color="grey49") + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y", ncol = 8) +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
 # expand_limits(y=0)+
  theme_bw(base_size = 14) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))
             
p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position
) + theme(legend.position="bottom")


#p

ggsave("Region_Violins.png", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)
ggsave("Region_Violins.pdf", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)

#MOs
ind<-which(violin_df$contrast=="MOs")
ind2<-which(violin_p_df$contrast=="MOs")

violin_df_ind<-violin_df[ind,]

p <- ggplot(violin_df_ind, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, position = position_dodge(0.8), color = NA) +
  geom_jitter(width=0.1, height=0, size = 0.5, color="grey49") +
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 14) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))
  

p <- p + ggprism::add_pvalue(
  violin_p_df[ind2,],
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position[ind2]
) + theme(legend.position="bottom")




ggsave("Region_Violins_MOs.png", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)
ggsave("Region_Violins_MOs.pdf", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)

#ACA
ind<-which(violin_df$contrast=="ACA")
ind2<-which(violin_p_df$contrast=="ACA")

violin_df_ind2<-violin_df[ind,]

p <- ggplot(violin_df_ind, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, color = NA) +
  geom_jitter(width=0.1, height=0, size = 1, color="grey49") + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 14) + 
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df[ind2,],
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position[ind2]
) + theme(legend.position="bottom")



p
ggsave("Region_Violins_ACA.png", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)
ggsave("Region_Violins_ACA.pdf", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)



#MOp
ind<-which(violin_df$contrast=="MOp")
ind2<-which(violin_p_df$contrast=="MOp")

violin_df_ind<-violin_df[ind,]


p <- ggplot(violin_df[ind,], 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, color = NA) +
  geom_jitter(width=0.1, height=0, size = 0.5, color="grey49") +
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 14) + 
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df[ind2,],
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position[ind2]
) + theme(legend.position="bottom")




ggsave("Region_Violins_MOp.png", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)
ggsave("Region_Violins_MOp.pdf", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)


```

Plot Cell Type genes

```{r}

cell_types<- c("Rorb", "Cux2", "Deptor", "Fezf2", "Lamp5", "Etv1", "Sulf1", "Nxph4", "Slc17a7", "Gad1", "Mbp","Snca" )

cell_types<- c("CUX2","EPHA6","RASGRF2","TMEM178B","CA10","FAM19A1","RFX3","KCTD16","KIRREL3","NTRK3","IL1RAPL1","LRRTM4","SYN3","CUX1","THSD7A","CAMK2A","PRKCA","PAM","SIPA1L1","TMEM108","RGS6","PTK2B","MAPK4","CACNA2D1","RORB","PTPRT","CNTN5","UNC5D","KCNC2","MEF2C","RUNX1T1","KCTD16","PTPRK","SHISA9","CDH9","TENM3","SLIT3","AK5","PCSK2","RGS12","TMEM132D","ERC2","KCTD16","ZNF804B","SORBS2","CUX1","LSAMP","NTNG2","MGAT5","UNC5C","GNG2","GALNT14","GNAO1","CACNA1E","CPNE5","CADPS2","SFMBT2","TENM4","SYNPR","CACNA1B","GNB4","KLHL1","GSG1L","SLC24A4","NR4A2","SYT1","GPD2","MBOAT2","ZNF804A","ENOX1","CPNE4","NBEA","SPOCK3","CHRM2","1-Mar","CACNA1C","DSCAML1","TMEM178B","GRIA4","TRPS1","TMEM163","OPRK1","ALCAM","TESC","PTPRU","ZEB2","HS3ST4","DLC1","TLE4","PITPNC1","PDZRN4","SLC35F1","SYT6","NRP1","MCTP1","GARNL3","PRKCB","ATP2B1","FOXP2","XKR4","KIAA1217","PCDH11X","1-Mar","FUT9","NRXN1","CDH10","NFIA","HS3ST4","EPHA5","DLC1","TNIK","TLE4","MDGA2","TENM1","SLC1A2","NFIB","PITPNC1","MCTP1","NR4A2","ACVR2A","RYR3","GAS7","SLC35F1","KCNMB4","PKP4","FRMPD4","MMP16","RAB3B","SLC8A1","GSG1L","VAT1L","GRIK2","ADRA1A","TMTC2","CDH20","FAM19A1","NRP1","NTNG1","LPP","SLIT2","SLC26A4","BCL11B","NEBL","ROBO2","ESRRG","PPARGC1A","NTM","RAB3C","FAM135B","STK39","SULF2","NFIB","ARL15","ANO4","EPHA6","DSCAML1","KCNN2","CHST8","SH3RF1","HTR4","HCN1","PEX5L","TCERG1L","POU3F1","TFDP2","FAM189A1","TOX2","GRIA3","ROBO1","FLT3","HDAC9","FRMD4A","RYR3","NAV1","CABP1","RAVER2","BCL6","DAB1","RAPGEF5","NEFH","GPRIN3","FAT3","SORCS2","FAM126A","NLGN1","TOX","FEZF2","NEFM","TRPC4","CTTNBP2","CDS1","GRAMD1B","PCGF5","ASAP1","WWOX","CACNA1H","TSHZ2","SORCS2","DAB1","KCNIP1","SPON1","GRM8","ETV1","ADCY2","KCNT2","VWC2L","PRR16","SATB1","OLFM3","C8orf34","SH3RF3","BCL11B","TLE4","MGAT4C","CAMK2D","CDH11","GRM3","PAK7","DCC","TOX","SASH1","CPNE4","CHSY3","STRBP","NETO2","SLC1A2","PCDH17","GRIA4")


cell_types<-tolower(cell_types)


cell_types<-tools::toTitleCase(cell_types)

ind<-c()
for (i in cell_types){
ind1<-which(results$Gene==i)
ind<-c(ind,ind1)
  }


results_cell<-results[ind,]

top_g <- c()


   for(i in unique(results_cell$contrast[ind])){
   vec<-which(results_cell$contrast==i) 
top_g <- c(top_g,
  results_cell[vec, 'Gene'][order(results_cell[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results_cell[vec, 'Gene'][order(results_cell[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)


highlight_top_g<-subset(results_cell, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

diff_exp3_cell<-ggplot(results_cell,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results_cell, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)


ggsave("DE_Cell_Markers.png", diff_exp3_cell,  width = 14, height = 7, units = "cm",path = de_plot_dir)
ggsave("DE_Cell_Markers.pdf", diff_exp3_cell,  width = 14, height = 7, units = "cm",path = de_plot_dir)





results_nocell<-results[-ind,]

top_g <- c()


   for(i in unique(results_nocell$contrast[ind])){
   vec<-which(results_nocell$contrast==i) 
top_g <- c(top_g,
  results_nocell[vec, 'Gene'][order(results_nocell[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results_nocell[vec, 'Gene'][order(results_nocell[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)


# Graph results
highlight_top_g<-subset(results_nocell, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

diff_exp3_cell<-ggplot(results_nocell,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results_nocell, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)


ggsave("DE_No_Cell_Markers.png", diff_exp3_cell,  width = 14, height = 7, units = "cm",path = de_plot_dir)
ggsave("DE_No_Cell_Markers.pdf", diff_exp3_cell,  width = 14, height = 7, units = "cm",path = de_plot_dir)

```









Run on whole dataset, no sub-setting
```{r}

# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$segment, c("pSyn", "NeuN"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))

vec<-unique(target_data@phenoData@data$layer)
pData(target_data)[["Layer"]] <- 
    factor(pData(target_data)$layer, c(vec))


vec<-unique(target_data@phenoData@data$region)
pData(target_data)[["Region"]] <- 
    factor(pData(target_data)$region, c(vec))


mixedOutmc <-
        mixedModelDE(target_data[,],
                     elt = "log_q",
                     modelFormula = ~ testRegion * Layer * Region + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results30<-formatLMMResults(mixedOutmc)
   
  #make gene nmae column
    results30$Gene<-results30$Feature  
   
 
  #remove negative probe
    ind<-which(results30$Gene=="NegProbe-WTX")
     results30<-results30[-ind,]
    
    
results30$Color[results30$P < 0.05 & results30$Estimate>0] <- "Enriched in pSyn P < 0.05"
results30$Color[results30$P < 0.05 & results30$Estimate<0] <- "Enriched in NeuN P < 0.05"

results30$Color[results30$FDR < 0.05 & results30$Estimate>0] <- "Enriched in pSyn FDR < 0.05"
results30$Color[results30$FDR < 0.05 & results30$Estimate<0] <- "Enriched in NeuN FDR < 0.05"


results30$Color[results30$FDR < 0.01 & results30$Estimate>0 ] <- "Enriched in pSyn FDR < 0.01"
results30$Color[results30$FDR < 0.01 & results30$Estimate<0 ] <- "Enriched in NeuN FDR < 0.01"

results30$Color[abs(results30$Estimate) < 0.5] <- "NS or FC < 0.5"


results30$Color <- factor(results30$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in pSyn FDR < 0.05", "Enriched in NeuN FDR < 0.05","Enriched in pSyn P < 0.05", "Enriched in NeuN P < 0.05", "Enriched in pSyn FDR < 0.01", "Enriched in NeuN FDR < 0.01" ))

# pick top genes for either side of volcano to label
# order genes for convenience:
results30$invert_P <- (-log10(results30$P)) * sign(results30$Estimate)

top_g <- c()



top_gene1<-results30[, 'Gene'][order(results30[, 'invert_P'], decreasing = TRUE)[1:15]]
top_gene2<-results30[, 'Gene'][order(results30[, 'invert_P'], decreasing = FALSE)[1:15]]
   
top_g<-c(top_gene1,top_gene2)

top_g <- unique(top_g)

highlight_top_g<-subset(results30, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

# Graph results30
diff_exp3<-ggplot(results30,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results30, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)

 diff_exp3
 
    
ggsave("volc_all.png", width = 10, height = 10, units = "cm",path = de_plot_dir)
ggsave("volc_all.pdf", width = 10, height = 10, units = "cm",path = de_plot_dir)
    
    
    
```




Violin Plotting for no downsampling

```{r}
#begin violin plotting
prop<-target_data@assayData$log_q
annots<-target_data@phenoData@data

violin_df <- cbind(annots %>% dplyr::select(eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

violin_p_df <- filter(results30, Feature %in% top_g)  
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
write_xlsx(violin_p_df,"AllDifferential_genes_pvalues_Mu.xlsx")
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

violin_df<-base::merge(violin_p_df, violin_df, by="Feature")

violin_df$FDR<-signif(violin_df$FDR, 3)

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, position = position_dodge(0.8), color = NA) +
  geom_jitter(width=0.1, height=0, size = 0.5, color="grey49") + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 14) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))
p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position
) + theme(legend.position="bottom")


p

ggsave("Violins_All.png", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)
ggsave("Violins_All.pdf", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)

```



Heatmap of Significant DE Genes for no subset
```{r}

sig_genes_psyn<-which(results30$Color=="Enriched in pSyn FDR < 0.05")

sig_genes_psyn2<-which(results30$Color=="Enriched in pSyn FDR < 0.01")

sig_genes_psyn<-c(sig_genes_psyn,sig_genes_psyn2)




sig_genes_NeuN<-which(results30$Color=="Enriched in NeuN FDR < 0.05")
sig_genes_NeuN2<-which(results30$Color=="Enriched in NeuN FDR < 0.01")
sig_genes_NeuN<-c(sig_genes_NeuN,sig_genes_NeuN2)


sig_genes<-c(sig_genes_NeuN,sig_genes_psyn)

sig_genes<-results30$Feature[sig_genes]
  
de_heat_dat_noSub<-target_data[sig_genes,]

annots<-de_heat_dat_noSub@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 


max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)

#color_pal<-colorRampPalette(c("black", "purple2", "yellow2"))(121)

color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(121))


#color_pal<-(colorRampPalette(c( "white","steelblue1", "firebrick1"))(121))


#color_pal<-(colorRampPalette(c("blue", "red"))(301))

heat_dat_maxmin<-de_heat_dat_noSub@assayData$log_q
  
# heat_dat_maxmin[which(heat_dat_maxmin>20)]<-20

 #heat_dat_maxmin[which(heat_dat_maxmin<1)]<-1

#color_pal<-viridis(350)


p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = FALSE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal
)




ggsave("Heatmap_FDR0.05_noSub.pdf",plot=p, width = 14, height = 14, units = "cm",path = de_plot_dir)


```




```{r}
write_xlsx(violin_p_df,"AllDifferentia_genes_pvalues_Mu.xlsx")

```



















Get heatmap that matched euler plot
```{r}

muind_pSyn<-which(results30$FDR < 0.05 & results30$Estimate>0)

length(muind_pSyn)

muind_neun<-which(results30$FDR < 0.05 & results30$Estimate<0)

length(muind_neun)

muind<-c(muind_neun,muind_pSyn)


mudat<-results30[muind,]

  
de_heat_dat_noSub<-target_data[muind,]

annots<-de_heat_dat_noSub@phenoData@data[,]
#annots<-annots[,c(3:11,13, 16,20)]
annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  layer= c("5"="greenyellow","6"="mediumorchid1","2/3"="lightcyan2", 	
"ND"="sandybrown")
 
) 


max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)

#color_pal<-colorRampPalette(c("black", "purple2", "yellow2"))(121)

color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(121))


#color_pal<-(colorRampPalette(c( "white","steelblue1", "firebrick1"))(121))


#color_pal<-(colorRampPalette(c("blue", "red"))(301))

heat_dat_maxmin<-de_heat_dat_noSub@assayData$log_q

# heat_dat_maxmin[which(heat_dat_maxmin>20)]<-20

 #heat_dat_maxmin[which(heat_dat_maxmin<1)]<-1

#color_pal<-viridis(350)


p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=1
)








ggsave("Heatmap_fdr005_noeffectsizefilter_noSub.pdf",plot=p, width = 14, height = 14, units = "cm",path = de_plot_dir)

```
























Get list of all significant genes for comparison to human data
```{r}
sig_genes_psyn_ind<-which(results30$FDR < 0.05 & results30$Estimate>0)

sig_genes_NeuN_ind<-which(results30$FDR < 0.05 & results30$Estimate<0)


sig_genes_psyn<-results30$Feature[sig_genes_psyn_ind]
saveRDS(sig_genes_psyn, "sig_genes_psyn_mu.rds")



sig_genes_NeuN<-results30$Feature[sig_genes_NeuN_ind]
saveRDS(sig_genes_NeuN,"sig_genes_NeuN_mu.rds")





```






violins of all conserved genes with human


```{r}

#begin violin plotting
prop<-target_data@assayData$log_q
annots<-target_data@phenoData@data

violin_df <- cbind(annots %>% dplyr::select(eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1), names_to = "Feature", values_to = "Expression")





conserved_genes<-readRDS("conservedGenes")

conserved_genes<-str_to_title(conserved_genes)

#get gene index for conserved genes
ind2<-c()
for(i in conserved_genes){
  ind<-which(results30$Gene==i)
  ind2<-c(ind,ind2)
}



violin_p_df <- results30[ind2,]




violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

violin_df<-base::merge(violin_p_df, violin_df, by="Feature")

violin_df$FDR<-signif(violin_df$FDR, 3)






p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, position = position_dodge(0.8), color = NA) +
  geom_jitter(width=0.1, height=0, size = 0.5, color="grey49") + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 14) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position
) + theme(legend.position="bottom")






conserved_plot_dir <-file.path(outdir, "Conserved_Genes", "Plots")

dir.create(conserved_plot_dir,recursive = TRUE)



ggsave("Violins_All_conserved_Mu.pdf", p,  width = 126, height = 76, units = "cm",path = conserved_plot_dir)

```



Get list of all conserved genes. estimates, and p values
```{r}

which(violin_p_df$Color=="NS or FC <0.5")


write_xlsx(violin_p_df,"conserved_genes_pvalues_Mu.xlsx")


```





















```{r}

library(readxl)

Gene_set_for_DGE_violin_plots <- read_excel("Gene set for DGE violin plots.xlsx")

genes<-Gene_set_for_DGE_violin_plots$Dnm1
genes[30]<-"Dnm1"


ind<-which(genes=="ATP5AP2")

genes[12]<-"Atp6ap2"




ind<-genes


#begin violin plotting
prop<-target_data@assayData$log_q[ind,]
annots<-target_data@phenoData@data

violin_df <- cbind(annots %>% dplyr::select(eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1), names_to = "Feature", values_to = "Expression")

ind2<-c()

for(i in rownames(prop)){
  print(i)
  ind<-which(results30$Gene==i)
  ind2<-c(ind2,ind)
}


ind2

violin_p_df <- results30[ind2,]




violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)



violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.15)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

violin_df<-base::merge(violin_p_df, violin_df, by="Feature")

violin_df$FDR<-signif(violin_df$FDR, 3)






p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, position = position_dodge(0.8), color = NA) +
  geom_jitter(width=0.12, height=0, size = 0.0000000000000001, color="grey49") + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 7) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.4,
  y.position = violin_p_df$y.position, bracket.size = 0.2
) + theme(legend.position="bottom")






conserved_plot_dir <-file.path(outdir, "Conserved_Genes", "Plots")

dir.create(PCA_plot_dir,recursive = TRUE)



ggsave("Violins_thirty_conserved_Mu.pdf", p,  width = 3.5, height = 4, units = "in",path = conserved_plot_dir)



p










```








Get genes of interest

```{r}
Example_DEGs <- read_excel("Example DEGs.xlsx")


```

```{r}
genes<-Example_DEGs$Mouse

ind<- genes

ind

#begin violin plotting
prop<-target_data@assayData$log_q[ind,]
annots<-target_data@phenoData@data

violin_df <- cbind(annots %>% dplyr::select(eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1), names_to = "Feature", values_to = "Expression")

ind2<-c()
for(i in rownames(prop)){
  ind<-which(results30$Gene==i)
  ind2<-c(ind2,ind)
}


ind2

violin_p_df <- results30[ind2,]




violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)



violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.15)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

violin_df<-base::merge(violin_p_df, violin_df, by="Feature")

violin_df$FDR<-signif(violin_df$FDR, 3)






p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2, position = position_dodge(0.8), color = NA) +
  geom_jitter(width=0.12, height=0, size = 0.0000000000000001, color="grey49") + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 7) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.4,
  y.position = violin_p_df$y.position, bracket.size = 0.2
) + theme(legend.position="bottom")






conserved_plot_dir <-file.path(outdir, "Conserved_Genes", "Plots")

dir.create(PCA_plot_dir,recursive = TRUE)



ggsave("Violins_Genes_of_Interest_Mu.pdf", p,  width = 3.5, height = 4, units = "in",path = conserved_plot_dir)



p
```












Lets look at zeng et al.s genesets

```{r}


Zeng_et_al_layer_selective_gene_table <- read_excel("Zeng et al. layer selective gene table.xlsx")

layer_markers_all<-Zeng_et_al_layer_selective_gene_table$Gene


layer_markers_all<-str_to_title(layer_markers_all)

#make dataset
NeuN_ind<-which(target_data@phenoData@data$segment=="NeuN")
marker_data<-target_data[,NeuN_ind]




genes_indnocon<-rownames(marker_data@assayData$log_q[,]) %in% layer_markers_all




annots<-marker_data@phenoData@data[,]

annots<-annots[,c(3,7,22)]

p<-pheatmap(marker_data@assayData$log_q[genes_indnocon,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)





ggsave("Heatmap_zengLayer_nocon_Hu.pdf",plot=p, width = 14, height = 14, units = "cm",path = de_plot_dir)







#now get sig genes

ind_allmarkers<-results30$Feature%in% layer_markers_all
results_markers<-results30[ind_allmarkers,]

heatmapdat_markers<-results_markers[results_markers$P < 0.05 & results_markers$Estimate>0|results_markers$P < 0.05 & results_markers$Estimate< 0,]

ind<-which(heatmapdat_markers$Color=="NS or FC < 0.5")
heatmapdat_markers<-heatmapdat_markers[-ind,]

genes_indallsig<-rownames(marker_data@assayData$log_q[,]) %in% heatmapdat_markers$Feature



p<-pheatmap(marker_data@assayData$log_q[genes_indallsig,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)
ggsave("Heatmap_zengLayer_signocon_Hu.pdf",plot=p, width = 14, height = 14, units = "cm",path = de_plot_dir)

```


Look at zeng genes with ordered cols by layer

```{r}


orddata<-marker_data@assayData$log_q[genes_indnocon,]



  
  
  

L5ind<-which(marker_data@phenoData@data$Layer=="5")
L6ind<-which(marker_data@phenoData@data$Layer=="6")

layallind<-c(L23ind,L5ind,L6ind)


orddata<-orddata[,layallind]


annots<-marker_data@phenoData@data[,]

annots<-annots[,c(3,7,22)]

annots<-annots[layallind,]


#convert to totitle
Zeng_et_al_layer_selective_gene_table$Gene<-str_to_title(Zeng_et_al_layer_selective_gene_table$Gene)

indall<-c()

for(i in Zeng_et_al_layer_selective_gene_table$Gene){
 ind1<-which(rownames(orddata)==i)
  indall<-c(indall,ind1)
}


orddata<-orddata[indall,]

p<-pheatmap(orddata,
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = FALSE,
         cluster_cols = FALSE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)



ggsave("Heatmap_zengLayer_ordered_Hu.pdf",plot=p, width = 14, height = 14, units = "cm",path = de_plot_dir)




#now get sig genes

ind_allmarkers<-results30$Feature%in% layer_markers_all
results_markers<-results30[ind_allmarkers,]

heatmapdat_markers<-results_markers[results_markers$P < 0.05 & results_markers$Estimate>0|results_markers$P < 0.05 & results_markers$Estimate< 0,]

ind<-which(heatmapdat_markers$Color=="NS or FC < 0.5")
heatmapdat_markers<-heatmapdat_markers[-ind,]

genes_indallsig<-rownames(marker_data@assayData$log_q[,]) %in% heatmapdat_markers$Feature

sig_gens_ind<-rownames(marker_data@assayData$log_q[genes_indallsig,])

ind2<-c()
for(i in sig_gens_ind){
  ind1<-which(rownames(orddata)==i)
  ind2<-c(ind1,ind2)
}


p<-pheatmap(orddata[ind2,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = FALSE,
         cluster_cols = FALSE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)

ggsave("Heatmap_zengLayer_ordered_HuSig.pdf",plot=p, width = 14, height = 14, units = "cm",path = de_plot_dir)

```












    RUN DE for L5 vs L6
```{r}
# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$segment, c("pSyn", "NeuN"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))

vec<-unique(target_data@phenoData@data$layer)
pData(target_data)[["Layer"]] <- 
    factor(pData(target_data)$layer, c(vec))
# run LMM:
# formula follows conventions defined by the lme4 package
results99 <- c()

    
    mixedOutmc <-
        mixedModelDE(target_data,
                     elt = "log_q",
                     modelFormula = ~ Layer * testRegion *region + (1|random),
                     groupVar = "Layer",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    
    

    # format results as data.frame
    results99 <-formatLMMResults(mixedOutmc)
    




results99$Gene<-results99$Feature


#remove neg probe from data
sub<-grep(".*NegP.*", results$Feature)
if(length(sub)>0){
results99<-results99[-sub,]}

# Categorize Results based on P-value & FDR for plotting
results99$Color<-c()

results99$Color[results99$P < 0.05 & results99$Estimate>0] <- "Enriched in L5 P < 0.05"
results99$Color[results99$P < 0.05 & results99$Estimate<0] <- "Enriched in L6 P < 0.05"

results99$Color[results99$FDR < 0.05 & results99$Estimate>0] <- "Enriched in L5 FDR < 0.05"
results99$Color[results99$FDR < 0.05 & results99$Estimate<0] <- "Enriched in L6 FDR < 0.05"


results99$Color[results99$FDR < 0.01 & results99$Estimate>0 ] <- "Enriched in L5 FDR < 0.01"
results99$Color[results99$FDR < 0.01 & results99$Estimate<0 ] <- "Enriched in L6 FDR < 0.01"


results99$Color[abs(results99$Estimate) < 0.5 | results99$P>0.05] <- "NS or FC < 0.5"


results99$Color <- factor(results99$Color,
                        levels = c("NS or FC < 0.5", "Enriched in L5 P < 0.05", "Enriched in L6 P < 0.05", "Enriched in L5 FDR < 0.05", "Enriched in L6 FDR < 0.05", "Enriched in L5 FDR < 0.01", "Enriched in L6 FDR < 0.01"  ))

# pick top genes for either side of volcano to label
# order genes for convenience:
results99$invert_P <- (-log10(results99$P)) * sign(results99$Estimate)

top_g <- c()


   
top_g <- c(top_g,
  results99[, 'Gene'][order(results99[, 'invert_P'], decreasing = TRUE)[1:10]],
  results99[, 'Gene'][order(results99[, 'invert_P'], decreasing = FALSE)[1:10]])
   


top_g <- unique(top_g)



highlight_top_g<-subset(results99, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")


diff_exp3<-ggplot(results99,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "Enriched in Layer 6<--log2(FC)--->Enriched in Layer 5",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in L5 FDR < 0.01`= "green",
                                  `Enriched in L5 FDR < 0.05` = "green3",
                                  `Enriched in L5 P < 0.05` = "greenyellow",
                                  `Enriched in L6 FDR < 0.01` = "mediumorchid1",
                                  `Enriched in L6 FDR < 0.05` = "mediumorchid4",
                                  `Enriched in L6 P < 0.05` = "thistle",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results99, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)
    
  



saveRDS(diff_exp3, file=file.path(de_data_dir, "Volc_L5vsL6.pdf"))
ggsave("Volc_L5vsL6.pdf", width = 14, height = 7, units = "cm",path = de_plot_dir)


diff_exp3






```




Get violins plots for each layer...not working yet

```{r}
#begin violin plotting
layer_sub<-which(target_data@phenoData@data$Layer==2)
prop<-target_data@assayData$log_q[,-layer_sub]
annots<-target_data@phenoData@data[-layer_sub,]
# 
# layer_sub<-which(annots$Layer==2)
# annots<-annots[-layer_sub,]
# 
# prop<-prop[,-layer]


violin_df<-c()

violin_df <- cbind(annots %>% dplyr::select(Layer),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

violin_p_df <- filter(results99, Feature %in% top_g) 
sub<-which(violin_p_df$Comparison==	"6 vs 5")
violin_p_df<-violin_p_df[sub,]
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

violin_df<-base::merge(violin_p_df, violin_df, by="Feature")

violin_df$FDR<-signif(violin_df$FDR, 3)

violin_p_df$Layer<-violin_p_df$Gene



p <- ggplot(violin_df, 
            aes(x=violin_df$Layer, y=Expression, fill=violin_df$Layer)) + 
 geom_violin(alpha=0.2, position = position_dodge(0.8), color = NA) +
  geom_jitter(width=0.1, height=0, size = 0.5, color="grey49") +
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y", ncol = 8) +
  labs(x = violin_df$Layer, y = "Expression (log q3 normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  #expand_limits(y=0)+
  theme_bw(base_size = 14) +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = violin_df$Layer)
+ theme(legend.position="bottom"))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 2.6,
  xmin= "group1", xmax= "group2", 
  y.position = violin_p_df$y.position
) 

p

ggsave("Violins_Layer.png", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)
ggsave("Violins_Layer.pdf", p,  width = 50, height = 30, units = "cm",path = de_plot_dir)

```








Run Each layer NeuN vs pSyn



```{r}
# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$segment, c("pSyn", "NeuN"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))

vec<-unique(target_data@phenoData@data$layer)
pData(target_data)[["Layer"]] <- 
    factor(pData(target_data)$layer, c(vec))
# run LMM:
# formula follows conventions defined by the lme4 package
results12 <- c()
for(status in c( "5", "6")) {
    ind <- which(pData(target_data)$Layer == status)
    mixedOutmc <-
        mixedModelDE(target_data[, ind],
                     elt = "log_q",
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results11 <-formatLMMResults(mixedOutmc)
    results11$contrast<-status
    
    results12 <- rbind(results11,results12)
    print(status)
}



results12$Gene<-results12$Feature


#remove neg probe from data
# sub<-grep(".*NegP.*", results$Feature)
# results12<-results12[-sub,]

# Categorize Results based on P-value & FDR for plotting
results$Color[results$P < 0.05 & results$Estimate>0] <- "Enriched in pSyn P < 0.05"
results$Color[results$P < 0.05 & results$Estimate<0] <- "Enriched in NeuN P < 0.05"

results$Color[results$FDR < 0.05 & results$Estimate>0] <- "Enriched in pSyn FDR < 0.05"
results$Color[results$FDR < 0.05 & results$Estimate<0] <- "Enriched in NeuN FDR < 0.05"


results$Color[results$FDR < 0.01 & results$Estimate>0 ] <- "Enriched in pSyn FDR < 0.01"
results$Color[results$FDR < 0.01 & results$Estimate<0 ] <- "Enriched in NeuN FDR < 0.01"

results$Color[abs(results$Estimate) < 0.5] <- "NS or FC < 0.5"



results$Color <- factor(results$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in pSyn FDR < 0.05", "Enriched in NeuN FDR < 0.05","Enriched in pSyn P < 0.05", "Enriched in NeuN P < 0.05", "Enriched in pSyn FDR < 0.01", "Enriched in NeuN FDR < 0.01" ))


# pick top genes for either side of volcano to label
# order genes for convenience:
results12$invert_P <- (-log10(results12$P)) * sign(results12$Estimate)
top_g <- c()


   for(i in unique(results12$contrast)){
   vec<-which(results12$contrast==i) 
top_g <- c(top_g,
  results12[vec, 'Gene'][order(results12[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results12[vec, 'Gene'][order(results12[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)


# Graph results
highlight_top_g<-subset(results12, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

diff_exp3<-ggplot(results12,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results12, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)


ggsave("Volc_by_Layer.png", diff_exp3_cell,  width = 14, height = 7, units = "cm",path = de_plot_dir)
ggsave("Volc_by_Layer.pdf", diff_exp3_cell,  width = 14, height = 7, units = "cm",path = de_plot_dir)

diff_exp3
```












Now we run DE by layer by region






Layer 5
```{r}
# run LMM:
# formula follows conventions defined by the lme4 package
ind<-which(target_data@phenoData@data$layer==5)
target_data2<-target_data[,ind]

results4 <- c()
for(status in c( "ACA", "MOs" , "MOp")) {
    ind <- which(pData(target_data2)$region == status)
    mixedOutmc <-
        mixedModelDE(target_data2[, ind],
                     elt = "log_q",
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results3 <-formatLMMResults(mixedOutmc)
    results3$contrast<-status
    
    results4 <- rbind(results4,results3)
    print(status)
}



results4$Gene<-results4$Feature


#remove neg probe from data
sub<-grep(".*NegP.*", results4$Feature)
results4<-results4[-sub,]

results4$Color<-c()

results4$Color[results4$P < 0.05 & results4$Estimate>0] <- "Enriched in pSyn P < 0.05"
results4$Color[results4$P < 0.05 & results4$Estimate<0] <- "Enriched in NeuN P < 0.05"

results4$Color[results4$FDR < 0.05 & results4$Estimate>0] <- "Enriched in pSyn FDR < 0.05"
results4$Color[results4$FDR < 0.05 & results4$Estimate<0] <- "Enriched in NeuN FDR < 0.05"


results4$Color[results4$FDR < 0.01 & results4$Estimate>0 ] <- "Enriched in pSyn FDR < 0.01"
results4$Color[results4$FDR < 0.01 & results4$Estimate<0 ] <- "Enriched in NeuN FDR < 0.01"

results4$Color[abs(results4$Estimate) < 0.5] <- "NS or FC < 0.5"


results4$Color <- factor(results4$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in pSyn FDR < 0.05", "Enriched in NeuN FDR < 0.05","Enriched in pSyn P < 0.05", "Enriched in NeuN P < 0.05", "Enriched in pSyn FDR < 0.01", "Enriched in NeuN FDR < 0.01" ))

# pick top genes for either side of volcano to label
# order genes for convenience:
results4$invert_P <- (-log10(results4$P)) * sign(results4$Estimate)

top_g <- c()

   for(i in unique(results4$contrast)){
   vec<-which(results4$contrast==i) 
top_g <- c(top_g,
  results4[vec, 'Gene'][order(results4[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results4[vec, 'Gene'][order(results4[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)

highlight_top_g<-subset(results4, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

# Graph results4
diff_exp3<-ggplot(results4,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results4, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)



saveRDS(diff_exp3, file=file.path(de_data_dir, "DE_Layer5.RDS"))
ggsave("Volc_Layer5.pdf", width = 14, height = 7, units = "cm",path = de_plot_dir)


diff_exp3










#Mos Only
diff_exp_mos<-ggplot(results4_Mos,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed") +
    geom_hline(yintercept = -log10(0.05), lty = "dashed") +
    geom_point() +
    labs(x = "Enriched in NeuN <- log2(FC) -> Enriched in pSyn",
         y = "Significance, -log10(P)",
         title = "Layer5",
         color = "Significance") +
    scale_color_manual(values = c(`FDR < 0.01` = "dodgerblue",
                                  `FDR < 0.05` = "lightblue",
                                  `P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                       guide = guide_legend(override.aes = list(size = 5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results4_Mos, Gene %in% top_g),
                    size = 6, point.padding = 0.15, color = "black",
                    min.segment.length = .1, box.padding = .2, lwd = 2,
                    max.overlaps = 50) +
    theme_bw(base_size = 25) +
    theme(legend.position = "bottom")+
  facet_wrap(~contrast)




saveRDS(diff_exp_mos, file=file.path(de_data_dir, "DE_Mos_Layer5.RDS"))
ggsave("diff_exp_Mos_Layer5.pdf", diff_exp_mos ,width = 50, height = 20, units = "cm",path = de_plot_dir)


diff_exp_mos





```















Layer6 
```{r}
# run LMM:
# formula follows conventions defined by the lme4 package
ind<-which(target_data@phenoData@data$layer==6)
target_data2<-target_data[,ind]


#remove mice with no psyn obs in MOs
mouse_sub<-c("IM 66", "IM 67", "IM 69", "IM 71", "IM 98", "IM 99")



results6 <- c()
for(status in c( "ACA", "MOs" , "MOp")) {
    ind <- which(pData(target_data2)$region == status)
    mixedOutmc <-
        mixedModelDE(target_data2[, ind],
                     elt = "log_q",
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results5 <-formatLMMResults(mixedOutmc)
    results5$contrast<-status
    
    results6 <- rbind(results6,results5)
    print(status)
}



results6$Gene<-results6$Feature


#remove neg probe from data
sub<-grep(".*NegP.*", results6$Feature)
results6<-results6[-sub,]

results6$Color<-c()

results6$Color[results6$P < 0.05 & results6$Estimate>0] <- "Enriched in pSyn P < 0.05"
results6$Color[results6$P < 0.05 & results6$Estimate<0] <- "Enriched in NeuN P < 0.05"

results6$Color[results6$FDR < 0.05 & results6$Estimate>0] <- "Enriched in pSyn FDR < 0.05"
results6$Color[results6$FDR < 0.05 & results6$Estimate<0] <- "Enriched in NeuN FDR < 0.05"


results6$Color[results6$FDR < 0.01 & results6$Estimate>0 ] <- "Enriched in pSyn FDR < 0.01"
results6$Color[results6$FDR < 0.01 & results6$Estimate<0 ] <- "Enriched in NeuN FDR < 0.01"

results6$Color[abs(results6$Estimate) < 0.5] <- "NS or FC < 0.5"


results6$Color <- factor(results6$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in pSyn FDR < 0.05", "Enriched in NeuN FDR < 0.05","Enriched in pSyn P < 0.05", "Enriched in NeuN P < 0.05", "Enriched in pSyn FDR < 0.01", "Enriched in NeuN FDR < 0.01" ))

# pick top genes for either side of volcano to label
# order genes for convenience:
results6$invert_P <- (-log10(results6$P)) * sign(results6$Estimate)

top_g <- c()

   for(i in unique(results6$contrast)){
   vec<-which(results6$contrast==i) 
top_g <- c(top_g,
  results6[vec, 'Gene'][order(results6[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results6[vec, 'Gene'][order(results6[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)

highlight_top_g<-subset(results6, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

# Graph results6
diff_exp3<-ggplot(results6,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "log2(FC)",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in pSyn FDR < 0.01`= "firebrick1",
                                  `Enriched in pSyn FDR < 0.05` = "rosybrown2",
                                  `Enriched in pSyn P < 0.05` = "rosybrown3",
                                  `Enriched in NeuN FDR < 0.01` = "dodgerblue2",
                                  `Enriched in NeuN FDR < 0.05` = "slategray2",
                                  `Enriched in NeuN P < 0.05` = "slategray3",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results6, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)


saveRDS(diff_exp, file=file.path(de_data_dir, "Volc_Layer6.RDS"))
ggsave("Volc_Layer6.pdf", diff_exp3 ,width = 14, height = 7, units = "cm",path = de_plot_dir)

diff_exp




#Mos Only
diff_exp_mos<-ggplot(results6_Mos,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed") +
    geom_hline(yintercept = -log10(0.05), lty = "dashed") +
    geom_point() +
    labs(x = "Enriched in NeuN <- log2(FC) -> Enriched in pSyn",
         y = "Significance, -log10(P)",
         title = "Layer5",
         color = "Significance") +
    scale_color_manual(values = c(`FDR < 0.01` = "dodgerblue",
                                  `FDR < 0.05` = "lightblue",
                                  `P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                       guide = guide_legend(override.aes = list(size = 5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results6_Mos, Gene %in% top_g),
                    size = 6, point.padding = 0.15, color = "black",
                    min.segment.length = .1, box.padding = .2, lwd = 2,
                    max.overlaps = 50) +
    theme_bw(base_size = 25) +
    theme(legend.position = "bottom")+
  facet_wrap(~contrast)



saveRDS(diff_exp_mos, file=file.path(de_data_dir, "DE_Mos_Layer6.RDS"))
ggsave("diff_exp_Mos_Layer6.pdf", diff_exp_mos ,width = 50, height = 20, units = "cm",path = de_plot_dir)

diff_exp_mos

```




















Layer 6 vs 5 Asyn +
```{r}
# run LMM:

# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$Layer, c("5", "6"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))










# formula follows conventions defined by the lme4 package
ind<-which(target_data@phenoData@data$segment=="pSyn")
target_data2<-target_data[,ind]

results8 <- c()
for(status in c( "ACA", "MOs" , "MOp")) {
    ind <- which(pData(target_data2)$region == status)
    mixedOutmc <-
        mixedModelDE(target_data2[, ind],
                     elt = "log_q",
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results7 <-formatLMMResults(mixedOutmc)
    results7$contrast<-status
    
    results8 <- rbind(results8,results7)
    print(status)
}



results8$Gene<-results8$Feature


#remove neg probe from data
sub<-grep(".*NegP.*", results8$Feature)
results8<-results8[-sub,]

results8$Color<-c()

results8$Color[results8$P < 0.05 & results8$Estimate>0] <- "Enriched in L5 P < 0.05"
results8$Color[results8$P < 0.05 & results8$Estimate<0] <- "Enriched in L6 P < 0.05"

results8$Color[results8$FDR < 0.05 & results8$Estimate>0] <- "Enriched in L5 FDR < 0.05"
results8$Color[results8$FDR < 0.05 & results8$Estimate<0] <- "Enriched in L6 FDR < 0.05"


results8$Color[results8$FDR < 0.01 & results8$Estimate>0 ] <- "Enriched in L5 FDR < 0.01"
results8$Color[results8$FDR < 0.01 & results8$Estimate<0 ] <- "Enriched in L6 FDR < 0.01"



results8$Color[abs(results8$Estimate) < 0.5] <- "NS or FC < 0.5"


results8$Color <- factor(results8$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in L5 FDR < 0.05", "Enriched in L6 FDR < 0.05","Enriched in L5 P < 0.05", "Enriched in L6 P < 0.05", "Enriched in L5 FDR < 0.01", "Enriched in L6 FDR < 0.01" ))

# pick top genes for either side of volcano to label
#downsmaple to regions
ind<-which(results8$contrast=="MOs")

results8_Mos<-results8[ind,]

# order genes for convenience:
results8$invert_P <- (-log10(results8$P)) * sign(results8$Estimate)

top_g <- c()


   for(i in unique(results8$contrast)){
   vec<-which(results8$contrast==i) 
top_g <- c(top_g,
  results8[vec, 'Gene'][order(results8[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results8[vec, 'Gene'][order(results8[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)


# Graph results
highlight_top_g<-subset(results8, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

diff_exp3<-ggplot(results8,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "Layer5 NeuN<- log2(FC) -> Layer6 NeuN",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in L5 FDR < 0.01`= "green",
                                  `Enriched in L5 FDR < 0.05` = "green3",
                                  `Enriched in L5 P < 0.05` = "greenyellow",
                                  `Enriched in L6 FDR < 0.01` = "mediumorchid1",
                                  `Enriched in L6 FDR < 0.05` = "mediumorchid4",
                                  `Enriched in L6 P < 0.05` = "thistle",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results8, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)



saveRDS(diff_exp3, file=file.path(de_data_dir, "Volc_Layer6vLayer5_pSyn.RDS"))
ggsave("Volc_Layer6vLayer5_pSyn.pdf", diff_exp3 ,width = 14, height = 7, units = "cm",path = de_plot_dir)

diff_exp

#Mos Only
diff_exp_mos<-ggplot(results8_Mos,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed") +
    geom_hline(yintercept = -log10(0.05), lty = "dashed") +
    geom_point() +
    labs(x = "Enriched in Layer 6 pSYn+ <- log2(FC) -> Enriched in Layer 5 pSyn+",
         y = "Significance, -log10(P)",
         title = "Layer5",
         color = "Significance") +
    scale_color_manual(values = c(`FDR < 0.01` = "dodgerblue",
                                  `FDR < 0.05` = "lightblue",
                                  `P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                       guide = guide_legend(override.aes = list(size = 5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results8_Mos, Gene %in% top_g),
                    size = 6, point.padding = 0.15, color = "black",
                    min.segment.length = .1, box.padding = .2, lwd = 2,
                    max.overlaps = 50) +
    theme_bw(base_size = 25) +
    theme(legend.position = "bottom")+
  facet_wrap(~contrast)



saveRDS(diff_exp_mos, file=file.path(de_data_dir, "DE_Mos_Layer6vLayer5_pSyn.RDS"))
ggsave("diff_exp_Mos_Layer6vLayer5_pSyn.pdf", diff_exp_mos ,width = 50, height = 20, units = "cm",path = de_plot_dir)


diff_exp_mos
```







Layer 6 vs 5 Neun
```{r}
# run LMM:

# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$Layer, c("5", "6"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))










# formula follows conventions defined by the lme4 package
ind<-which(target_data@phenoData@data$segment=="NeuN")
target_data2<-target_data[,ind]

results9 <- c()
for(status in c( "ACA", "MOs" , "MOp")) {
    ind <- which(pData(target_data2)$region == status)
    mixedOutmc <-
        mixedModelDE(target_data2[, ind],
                     elt = "log_q",
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results10 <-formatLMMResults(mixedOutmc)
    results10$contrast<-status
    
    results9 <- rbind(results9,results10)
    print(status)
}



results9$Gene<-results9$Feature


#remove neg probe from data
sub<-grep(".*NegP.*", results9$Feature)
results9_Mos<-results9[-sub,]


results9$Color<-c()

results9$Color[results9$P < 0.05 & results9$Estimate>0] <- "Enriched in L5 P < 0.05"
results9$Color[results9$P < 0.05 & results9$Estimate<0] <- "Enriched in L6 P < 0.05"

results9$Color[results9$FDR < 0.05 & results9$Estimate>0] <- "Enriched in L5 FDR < 0.05"
results9$Color[results9$FDR < 0.05 & results9$Estimate<0] <- "Enriched in L6 FDR < 0.05"


results9$Color[results9$FDR < 0.01 & results9$Estimate>0 ] <- "Enriched in L5 FDR < 0.01"
results9$Color[results9$FDR < 0.01 & results9$Estimate<0 ] <- "Enriched in L6 FDR < 0.01"



results9$Color[abs(results9$Estimate) < 0.5] <- "NS or FC < 0.5"


results9$Color <- factor(results9$Color,
                        levels = c("NS or FC < 0.5" , "Enriched in L5 FDR < 0.05", "Enriched in L6 FDR < 0.05","Enriched in L5 P < 0.05", "Enriched in L6 P < 0.05", "Enriched in L5 FDR < 0.01", "Enriched in L6 FDR < 0.01" ))

# pick top genes for either side of volcano to label
#downsmaple to regions
ind<-which(results9$contrast=="MOs")

results9_Mos<-results9[ind,]

# order genes for convenience:
results9$invert_P <- (-log10(results9$P)) * sign(results9$Estimate)

top_g <- c()


   for(i in unique(results9$contrast)){
   vec<-which(results9$contrast==i) 
top_g <- c(top_g,
  results9[vec, 'Gene'][order(results9[vec, 'invert_P'], decreasing = TRUE)[1:5]],
  results9[vec, 'Gene'][order(results9[vec, 'invert_P'], decreasing = FALSE)[1:5]])
   }


top_g <- unique(top_g)


# Graph results
highlight_top_g<-subset(results9, Gene %in% top_g & P<0.05 &  Color != "NS or FC < 0.5")

diff_exp3<-ggplot(results9,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed", size=0.2) +
    geom_hline(yintercept = -log10(0.05), lty = "dashed", size=0.2) +
    geom_point(size=0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001, alpha=0.3) +
    labs(x = "Layer5 NeuN<- log2(FC) -> Layer6 NeuN",
         y = "Significance, -log10(P)",
         color = "Key") +
    scale_color_manual(values = c(`Enriched in L5 FDR < 0.01`= "green",
                                  `Enriched in L5 FDR < 0.05` = "green3",
                                  `Enriched in L5 P < 0.05` = "greenyellow",
                                  `Enriched in L6 FDR < 0.01` = "mediumorchid1",
                                  `Enriched in L6 FDR < 0.05` = "mediumorchid4",
                                  `Enriched in L6 P < 0.05` = "thistle",
                                  #`P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                                  guide = guide_legend(override.aes = list(size = 0.5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results9, Gene %in% top_g & P<0.05 & Color != "NS or FC < 0.5"),
                    size = 1.5, point.padding = 0.1, color = "black",
                    min.segment.length = .3, box.padding = .1, lwd = .2,
                    max.overlaps = 50, segment.size=0.05, force = 10, max.time = 3) +
    theme_bw(base_size = 6) +
  theme(axis.line = element_line(color='black'),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
    theme(legend.position = "bottom",
          legend.key.size= unit(0.00001, 'cm'),
          legend.title = element_text(size = 2),
          legend.text = element_text(size=2),
          legend.key.height = unit(0.3, 'cm'),
          legend.key.width = unit(3, 'cm')) +
  geom_point(data=highlight_top_g, alpha=0.9, size=0.5)+
    facet_wrap(~contrast, nrow = 1, ncol = 3)




diff_exp3

saveRDS(diff_exp3, file=file.path(de_data_dir, "DE_Layer6vLayer5_NeuN.RDS"))
ggsave("Volc_Layer6vLayer5_NeuN.pdf", diff_exp3 ,width = 14, height = 7, units = "cm",path = de_plot_dir)



#Mos Only
diff_exp_mos<-ggplot(results9_Mos,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Gene)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed") +
    geom_hline(yintercept = -log10(0.05), lty = "dashed") +
    geom_point() +
    labs(x = "Enriched in Layer 6 pSYn+ <- log2(FC) -> Enriched in Layer 5 pSyn+",
         y = "Significance, -log10(P)",
         title = "Layer5",
         color = "Significance") +
    scale_color_manual(values = c(`FDR < 0.01` = "dodgerblue",
                                  `FDR < 0.05` = "lightblue",
                                  `P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                       guide = guide_legend(override.aes = list(size = 5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(results9_Mos, Gene %in% top_g),
                    size = 6, point.padding = 0.15, color = "black",
                    min.segment.length = .1, box.padding = .2, lwd = 2,
                    max.overlaps = 50) +
    theme_bw(base_size = 25) +
    theme(legend.position = "bottom")+
  facet_wrap(~contrast)



saveRDS(diff_exp_mos, file=file.path(de_data_dir, "DE_Mos_Layer6vLayer5_NeuN.RDS"))
ggsave("diff_exp_Mos_Layer6vLayer5_NeuN.pdf", diff_exp_mos ,width = 50, height = 20, units = "cm",path = de_plot_dir)






diff_exp_mos
```





Heatmap of Layer specific genes
```{r}

# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$Layer, c("5", "6"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))










# formula follows conventions defined by the lme4 package
ind<-which(target_data@phenoData@data$segment=="NeuN")
target_data2<-target_data[,ind]



    mixedOutmc <-
        mixedModelDE(target_data2,
                     elt = "log_q",
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)
    
    # format results as data.frame
    results_neunlayer <-formatLMMResults(mixedOutmc)
    
   
   



results_neunlayer$Gene<-results_neunlayer$Feature

```

```{r}



l5genesdat<-results_neunlayer[results_neunlayer$FDR < 0.01 & results_neunlayer$Estimate>0,]
l6genesdat<-results_neunlayer[results_neunlayer$FDR < 0.01 & results_neunlayer$Estimate<0,]


l5genesdat<-l5genesdat[l5genesdat$Estimate>0.5,]
l6genesdat<-l6genesdat[l6genesdat$Estimate< -0.5,]


l5genes<-l5genesdat$Gene
l6genes<-l6genesdat$Gene
l5l6genes<-c(l5genes,l6genes)


genes_ind<-rownames(target_data2@assayData$log_q) %in% l5l6genes

annots<-target_data2@phenoData@data
annots<-annots[,c(3,7,22)]


p<-pheatmap(target_data2@assayData$log_q[genes_ind,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)




#now leets try the Human DE genes

l56geneshu<-c("Syt2", "Sncg", "Rorb","Vat1l", "Cacna1e", "Synpr", "B3galt2","Tle4", "Rprm", "Kcnip2")
genes_ind<-rownames(target_data2@assayData$log_q) %in% l56geneshu

p<-pheatmap(target_data2@assayData$log_q[genes_ind,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)

```










































Begin Decon


ACA

```{r}
#downsample the data so all tech replicates are averaged
# vars<- c("segment", "region")
# 
# 
# norm<-downsample(target_data,"q_norm" , vars)

#only run once

```






```{r}

norm<-target_data@assayData$q_norm

nuc_count<-target_data@phenoData@data$AOINucleiCount
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

bg = derive_GeoMx_background(norm = target_data@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")



```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/ACA_Profile_.csv")

#remove non nuerons
profile_matrix<-profile_matrix[,-c(20:23)]
rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname


```



```{r}

ind<-which(target_data@phenoData@data$region=="ACA")
norm<-norm[,ind]
bg<-bg[,ind]
nuc_count<-nuc_count[ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}


pData(target_data)$testRegion <- 
    factor(pData(target_data)$segment, c("pSyn", "NeuN"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))

vec<-unique(target_data@phenoData@data$layer)
pData(target_data)[["Layer"]] <- 
    factor(pData(target_data)$layer, c(vec))


annots<-target_data@phenoData@data

ind<-which(annots$region=="ACA")

annots<-annots[ind,]

annots<-annots[,c(7,3,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)
ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, max , 0.01) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*100)
)



decon_plot_dir <-file.path(outdir, "Decon", "Plots")
decon_data_dir <- file.path(outdir, "Decon", "Data")
dir.create(decon_plot_dir,recursive = TRUE)
dir.create(decon_data_dir, recursive = TRUE)


saveRDS(p, file=file.path(decon_data_dir, "Decon_ACA.RDS"))
 ggsave("Decon_ACA.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)

 
 
```
 
 
 
 Run DE on decon results to get significance of differences
```{r}

# convert test variables to factors
pData(target_data)$testRegion <- 
    factor(pData(target_data)$segment, c("pSyn", "NeuN"))

vec<-unique(target_data@phenoData@data$mouse)
pData(target_data)[["random"]] <- 
    factor(pData(target_data)$mouse, c(vec))

vec<-unique(target_data@phenoData@data$layer)
pData(target_data)[["Layer"]] <- 
    factor(pData(target_data)$layer, c(vec))





annots<-target_data@phenoData@data
ind<-which(target_data@phenoData@data$region=="ACA")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores = parallel::detectCores()-1,
      multiCore = TRUE)

  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
```

Get violins
```{r}




  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

violin_df$contrast<-violin_df$region
  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_ACA_Violins.RDS"))
 ggsave("Decon_ACA_Violins.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)

p
```







MOS




```{r}
#downsample the data so all tech replicates are averaged
# vars<- c("segment", "region")
# 
# 
# norm<-downsample(target_data,"q_norm" , vars)

norm<-target_data@assayData$q_norm

nuc_count<-target_data@phenoData@data$AOINucleiCount
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = target_data@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/Mos_Profile.csv")

#remove non nuerons
profile_matrix<-profile_matrix[,-c(17,18)]

rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname


```



```{r}

ind<-which(target_data@phenoData@data$region=="MOs")
norm<-norm[,ind]
bg<-bg[,ind]
nuc_count<-nuc_count[ind]
```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-target_data@phenoData@data
ind<-which(target_data@phenoData@data$region=="MOs")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_MOs.RDS"))
 ggsave("Decon_MOs.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


```

 Run DE on decon results to get significance of differences
```{r}
annots<-target_data@phenoData@data
ind<-which(target_data@phenoData@data$region=="MOs")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_MOs_Violins.RDS"))
 ggsave("Decon_MOs_Violins.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


p
```



MOP





```{r}
#downsample the data so all tech replicates are averaged
# vars<- c("segment", "region")
# 
# 
# norm<-downsample(target_data,"q_norm" , vars)

norm<-target_data@assayData$q_norm

nuc_count<-target_data@phenoData@data$AOINucleiCount
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

bg = derive_GeoMx_background(norm = target_data@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/Mop_Profile_Final.csv")
profile_matrix<-profile_matrix[,-c(14,19 ,17,21:22)]
rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname

#remove non nuerons

```



```{r}

ind<-which(target_data@phenoData@data$region=="MOp")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-target_data@phenoData@data
ind<-which(target_data@phenoData@data$region=="MOp")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
          

p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, max , 0.01) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*100)
)





saveRDS(p, file=file.path(decon_data_dir, "Decon_MOp.RDS"))
 ggsave("Decon_MOp.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


```


 Run DE on decon results to get significance of differences
```{r}
annots<-target_data@phenoData@data
ind<-which(target_data@phenoData@data$region=="MOp")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_MOp_Violins.RDS"))
 ggsave("Decon_MOp_Violins.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


p
```










Layer 5 MOS

```{r}
#subset for layer 5
ind<-which(target_data@phenoData@data$Layer=="5")


L5target<-target_data[,ind]


norm<-L5target@assayData$q_norm

nuc_count<-L5target@phenoData@data$AOINucleiCount[ind]
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = L5target@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/Mos_Profile.csv")

#remove non nuerons
profile_matrix<-profile_matrix[,-c(17,18)]

rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname


```



```{r}

ind<-which(L5target@phenoData@data$region=="MOs")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOs")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_MOs_layer5.RDS"))
 ggsave("Decon_MOs_layer5.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


```

 Run DE on decon results to get significance of differences
```{r}






annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOs")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y",nrow=4, ncol=4) +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))
p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_MOs_Violins_layer5.RDS"))
 ggsave("Decon_MOs_Violins_layer5.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


p
```










Layer 6 MOS

```{r}
#subset for layer 5
ind<-which(target_data@phenoData@data$Layer=="6")


L5target<-target_data[,ind]


norm<-L5target@assayData$q_norm

nuc_count<-L5target@phenoData@data$AOINucleiCount[ind]
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = L5target@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/Mos_Profile.csv")

#remove non nuerons
profile_matrix<-profile_matrix[,-c(17,18)]

rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname


```



```{r}

ind<-which(L5target@phenoData@data$region=="MOs")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOs")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_MOs_layer6.RDS"))
 ggsave("Decon_MOs_layer6.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


```

 Run DE on decon results to get significance of differences
```{r}






annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOs")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_MOs_Violins_layer6.RDS"))
 ggsave("Decon_MOs_Violins_layer6.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


p
```








Layer 5 ACA

```{r}
#subset for layer 5
ind<-which(target_data@phenoData@data$Layer=="5")


L5target<-target_data[,ind]


norm<-L5target@assayData$q_norm

nuc_count<-L5target@phenoData@data$AOINucleiCount[ind]
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = L5target@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/ACA_Profile_.csv")

#remove non nuerons
profile_matrix<-profile_matrix[,-c(20:23)]
rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname


```



```{r}

ind<-which(L5target@phenoData@data$region=="ACA")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="ACA")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_ACA_layer5.RDS"))
 ggsave("Decon_ACA_layer5.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


```

 Run DE on decon results to get significance of differences
```{r}






annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="ACA")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)


saveRDS(p, file=file.path(decon_data_dir, "Decon_ACA_Violins_layer5.RDS"))
 ggsave("Decon_ACA_Violins_layer5.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


p
```










Layer 6 ACA

```{r}
#subset for layer 5
ind<-which(target_data@phenoData@data$Layer=="6")


L5target<-target_data[,ind]


norm<-L5target@assayData$q_norm

nuc_count<-L5target@phenoData@data$AOINucleiCount[ind]
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = L5target@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/ACA_Profile_.csv")

#remove non nuerons
profile_matrix<-profile_matrix[,-c(20:23)]
rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname


```



```{r}

ind<-which(L5target@phenoData@data$region=="ACA")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="ACA")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_ACA_layer6.RDS"))
 ggsave("Decon_ACA_layer6.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


```


 Run DE on decon results to get significance of differences
```{r}






annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="ACA")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
 geom_violin(alpha=0.2, lwd=0.1, position = position_dodge(0.8), color = NA)+ 
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) +
  geom_jitter(width=0.2, height=0, size = 0.3, color="gray49") + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 6) +
  theme(legend.position = "bottom") +
  theme( plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())+
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 1.8, bracket.size = 0.3,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_ACA_Violins_layer6.RDS"))
 ggsave("Decon_ACA_Violins_layer6.pdf", p,  width = 15, height = 10, units = "cm",path = decon_plot_dir)


p
```








Layer 5 MOp

```{r}
#subset for layer 5
ind<-which(target_data@phenoData@data$Layer=="5")


L5target<-target_data[,ind]


norm<-L5target@assayData$q_norm

nuc_count<-L5target@phenoData@data$AOINucleiCount[ind]
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = L5target@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/Mop_Profile_Final.csv")
profile_matrix<-profile_matrix[,-c(14,19 ,17,21:22)]
rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname

#remove non nuerons

```


```{r}

ind<-which(L5target@phenoData@data$region=="MOp")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOp")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_MOp_layer5.RDS"))
 ggsave("Decon_MOp_layer5.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


```

 Run DE on decon results to get significance of differences
```{r}






annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOp")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=segment, y=Expression, fill=segment)) + 
  geom_violin(alpha=0.2) +
  # geom_jitter(aes(colour=slide), width=0.25, height=0, size = 1.5) + 
  geom_jitter(width=0.25, height=0, size = 1.5) + 
 scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  xlab(eval(segment)) +
  ylab("Cell Abundance (Proportion)") + 
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.2))) +
  theme_bw(base_size = 14) +
  theme(legend.position = "bottom") + 
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position + violin_p_df$y.position*0.2
)

saveRDS(p, file=file.path(decon_data_dir, "Decon_MOp_Violins_layer5.RDS"))
 ggsave("Decon_MOp_Violins_layer5.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


p
```










Layer 6 ACA

```{r}
#subset for layer 5
ind<-which(target_data@phenoData@data$Layer=="6")


L5target<-target_data[,ind]


norm<-L5target@assayData$q_norm

nuc_count<-L5target@phenoData@data$AOINucleiCount[ind]
# var_meta<- c("slide","layer", "scan", "segment", "area", "mouse", "region" )
# 
# test <- pData(target_data)[,var_meta]
# rows <- paste(test$segment,test$Case,test$Layer,sep="_")
# sub1 <- match(colnames(A), rows)
# B<- pData(target_data)[sub1,var_meta]

```



```{r}

#Find the background
bg = derive_GeoMx_background(norm = L5target@assayData$q_norm,   #need to use the normalized data (e.g. the targetdemo data)
                             probepool = fData(target_data)$Module,   
negnames = "NegProbe-WTX")




```





```{r}
#import profle marix
profile_matrix<-read_csv("~/Henderson_Lab/Deconvolution/A_Thomas_GeoMx_05_31_22/Profile_Matrix_Region_Specific_Allen/Mop_Profile_Final.csv")
profile_matrix<-profile_matrix[,-c(14,19 ,17,21:22)]
rowname<-profile_matrix$...1
profile_matrix<-profile_matrix[,-1]
profile_matrix<-as.matrix(profile_matrix)
rownames(profile_matrix)<-rowname

#remove non nuerons

```



```{r}

ind<-which(L5target@phenoData@data$region=="MOp")
norm<-norm[,ind]
bg<-bg[,ind]

```

```{r}
res = spatialdecon(norm = norm,#normalized data matrix       #background use bg2
                   bg=bg,
                   X = profile_matrix,   #profile matrix
                   align_genes = TRUE,
                   cell_counts = nuc_count)


```



```{r}

annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOp")
annots<-annots[ind,]
annots<-annots[,c(3,7,22)]

# rownames(annots)<- colnames(res$prop_of_all)
# 
#annots<-t(annots)
# annots<-as.matrix(annots)

ind<-which(rowMeans(res$prop_of_all)==0)
max<-c()

max<-max(res$prop_of_all)
min<-min(res$prop_of_all)


ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1")
) 
            


p<-pheatmap(t(res$prop_of_all[-ind,]), 
scale = "none", 
         show_rownames = FALSE, show_colnames = TRUE,
         border_color = NA,
         clustering_method = "average",
         cluster_rows = TRUE,
        cluster_cols = TRUE,
         clustering_distance_cols = "correlation",
  annotation_row = annots,
annotation_colors = ann_colors,
 breaks = seq(min, 1 , 0.02) ,
         color = colorRampPalette(c("black", "purple2", "yellow2"))(max*50+1)
)



saveRDS(p, file=file.path(decon_data_dir, "Decon_MOp_layer6.RDS"))
 ggsave("Decon_MOp_layer6.pdf", p,  width = 30, height = 20, units = "cm",path = decon_plot_dir)


```


 Run DE on decon results to get significance of differences
```{r}






annots<-L5target@phenoData@data
ind<-which(L5target@phenoData@data$region=="MOp")
annots<-annots[ind,]



prop<-res$prop_of_all
prop <- prop[apply(prop, 1, sum)!=0,]
 
colnames(prop)<-annots$segment
 
melt_prop<-melt(prop)
 

da_contrast1<-mixedDE(object=prop,
      pdat=annots,
      modelFormula = ~ testRegion + (1|random),
      groupVar="testRegion",
      nCores=parallel::detectCores()-1,
      multiCore=TRUE
    )
  da_contrast1 <- formatLMMResults(da_contrast1)
  
  
  
  
  violin_df <- cbind(annots %>% dplyr::select(`slide`, eval(segment)),
                  t(prop))
violin_df <- violin_df %>% tidyr::pivot_longer(cols=-c(1,2), names_to = "Feature", values_to = "Expression")

  
violin_p_df <- da_contrast1
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$P <- signif(violin_p_df$P, 3)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=(max(Expression)*1.1)) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

# Cap at 16 unless top_features is >16
if(nrow(da_contrast1)>16){
  violin_p_df <- violin_p_df %>% filter(Feature %in% label_da_contrast1$all) %>% arrange(FDR) %>% head(16)
  violin_df <- violin_df %>% filter(Feature %in% violin_p_df$Feature)
}

# Sort plot by P-value
violin_p_df$Feature <- factor(violin_p_df$Feature, levels=arrange(violin_p_df, FDR)$Feature, order=TRUE)
violin_df$Feature <- factor(violin_df$Feature, levels=levels(violin_p_df$Feature), order=TRUE)
violin_df <- violin_df %>% as.data.frame(arrange(violin_df, Feature))
 

p <- ggplot(violin_df, 
            aes(x=contrast_factor, y=Enrichment, fill=segment)) + 
  geom_violin(alpha=0.2) +
  geom_jitter(width=0.1, height=0, size = 1) + 
  scale_fill_manual(values = c("blue", "red", "grey")) +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(segment), y = "Expression (normalized counts)") +
  scale_y_continuous(expand = expansion(mult = 0.2)) +
  theme_bw(base_size = 14) + 
  guides(fill=guide_legend(title = eval(segment)))

p <- p + ggprism::add_pvalue(
  violin_p_df[ind2,],
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position[ind2]
) + theme(legend.position="bottom")



p
```



























Run GSEA

To run GSEA you must have added a geneset to the working directory of this analysis. Make sure you downlaod the SYMBOL files





```{r}
GSEA_plot_dir <-file.path(datadir, "Gene_Set")
```


```{r}
gmt<-read.gmt("C:/Users/Thomas.Goralski/Documents/Henderson_Lab/Deconvolution/Final_GeoMX_Pipeline/Gene_Set/kegg_brite_all_ENTREZID.gmt")


```




```{r}
# point the function to a folder containing GMT files or to a gene set list


geneSet_data <- geneSetAnalysis(object = target_data,
                                 elt = "log_q",
                                 geneSet = GSEA_plot_dir,
                                 convertFrom = "ENTREZID",
                                 species = "Mm",
                                 minSize = 5,
                                 maxSize = 500)

 #saveRDS(geneSet_data, file.path(object_dir, "geneSetObject.RDS"))

# to view information about the gene sets scored, view the phenoData using:
#  pData(geneSet_data)

# check for gene sets of interest being captured in the analysis:

```


```{r}



 geneSetDE_contrast1<-
        mixedModelDE(geneSet_data,
                     modelFormula = ~ testRegion + (1|random),
                     groupVar = "testRegion",
                     nCores = parallel::detectCores()-1,
                     multiCore = TRUE)

geneSet_results_contrast1 <- formatLMMResults(geneSetDE_contrast1)


strings <- do.call(rbind, strsplit(unique(geneSet_results_contrast1$Comparison), split=" vs "))


# parameters for selecting top sets
volcano_fc_thr <- 0.5
volcano_p_thr <- 0.05
fdr_threshold1 <- 0.2
fc_threshold1 <- 0.5
fdr_threshold2 <- 0.2
fc_threshold2 <- 0.50
fdr_threshold3 <- 0.2
fc_threshold3 <- 0.50
fdr_threshold4 <- 0.2
fc_threshold4 <- 0.50

top_feat_gs_contrast1 <- getTopFeatures(geneSet_results_contrast1, n_features = 8,
                            est_thr = fc_threshold1, fdr_thr = fdr_threshold2)
```




```{r}

geneSet_results_contrast1$Beta<-abs(geneSet_results_contrast1$Estimate)

geneSet_results_contrast2<-geneSet_results_contrast1[which(geneSet_results_contrast1$FDR<0.001),]
geneSet_results_contrast2$Beta<-abs(geneSet_results_contrast2$Estimate)
color<-c()
for (i in geneSet_results_contrast2$Estimate){
  if (i<0)
    color1<- "#0000FF"
  
  if(i>0)
    color1<- "#FF0000"
  
  color<-c(color, color1)
}


geneSet_results_contrast2$color<-color


p<-ggplot(geneSet_results_contrast1[,], # you can replace the numbers to the row number of pathway of your interest
             aes(x =FDR, y =Feature)) + 
             geom_point(aes(size = Beta, color=Estimate)) +
  scale_color_gradient(low = "blue", high = "red")+
             theme_bw(base_size = 12) +
             theme(axis.text = element_text(size=9))+
             ylab(NULL) +
             ggtitle("GeneSet Enrichment Analysis")

p
```







```{r}

top_feat<- getTopFeatures(geneSet_results_contrast1, n_features = 30,
                            est_thr = 0.5, fdr_thr = 0.001)
saveRDS(top_feat,"top_genesets_mouse")

geneSet_results_contrast1$Color[geneSet_results_contrast1$P > 0.05] <- "NS or FC < 0.5"
geneSet_results_contrast1$Color[geneSet_results_contrast1$P < 0.05] <- "P < 0.05"
geneSet_results_contrast1$Color[geneSet_results_contrast1$FDR < 0.05] <- "FDR < 0.05"
geneSet_results_contrast1$Color[geneSet_results_contrast1$FDR < 0.01] <- "FDR < 0.01"
geneSet_results_contrast1$Color[abs(geneSet_results_contrast1$Estimate) < 0.5] <- "NS or FC < 0.5"
geneSet_results_contrast1$Color <- factor(geneSet_results_contrast1$Color,
                        levels = c("NS or FC < 0.5", "P < 0.05",
                                   "FDR < 0.05", "FDR < 0.01"))

gsea<-ggplot(geneSet_results_contrast1,
       aes(x = Estimate, y = -log10(`P`),
           color = Color, label = Feature)) +
    geom_vline(xintercept = c(0.5, -0.5), lty = "dashed") +
    geom_hline(yintercept = -log10(0.05), lty = "dashed") +
    geom_point() +
    labs(x = "Enriched in pSYn+ <- log2(FC) -> Enriched in NeuN",
         y = "Significance, -log10(P)",
         title = "Vulnerable Vs. Resilient Neurons",
         color = "Significance") +
    scale_color_manual(values = c(`FDR < 0.01` = "dodgerblue",
                                  `FDR < 0.05` = "lightblue",
                                  `P < 0.05` = "orange2",
                                  `NS or FC < 0.5` = "gray"),
                       guide = guide_legend(override.aes = list(size = 5))) +
    scale_y_continuous(expand = expansion(mult = c(0,0.05))) +
    geom_text_repel(data = subset(geneSet_results_contrast1, Feature %in% top_feat$all),
                    size = 4, point.padding = 0.15, color = "black",
                    min.segment.length = .1, box.padding = .2, lwd = 2,
                    max.overlaps = 50) +
    theme_bw(base_size = 25) +
    theme(legend.position = "bottom")




gsea



gsea_plot_dir <-file.path(outdir, "GSEA", "Plots")
gsea_data_dir <- file.path(outdir, "GSEA", "Data")
dir.create(gsea_plot_dir,recursive = TRUE)
dir.create(gsea_data_dir, recursive = TRUE) 



saveRDS(gsea, file=file.path(gsea_data_dir, "gsea_volcano.RDS"))
 ggsave("gsea_volcano.png",plot=gsea,  width = 30, height = 20, units = "cm",path = gsea_plot_dir)


```








```{r}
annots<-de_heat_dat_noSub@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 

heatmap_geneset<-geneSet_data@assayData$ssgsea


label_ge1 <- unique(top_feat$all)




#search for keywork in geneset here
grep("Ubiquitin",geneSet_data@featureData@data$GeneSet)

#make in index using that keyword
ind<-grep("Ubiquitin", geneSet_data@featureData@data$GeneSet)

#list results of keyword search
geneSet_data@featureData@data$GeneSet[ind]

#place desired full geneset name in the quotations to get geneset index
geneset_sub<-which(geneSet_data@featureData@data$GeneSet=="Ubiquitin mediated proteolysis")


#this index has the indices for each gene-sets we are interested in
geneset_ind<-c(geneset_sub,geneset_ind)



p<-pheatmap(heatmap_geneset[geneset_ind,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)

ggsave("Heatmap_Genesets_Of_Interest.pdf",plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)
```







```{r}
hu_gsi<-readRDS("Human_genesets_of_interest")

geneset_ind<-c()
for (i in hu_gsi){
  geneset_sub<-which(geneSet_data@featureData@data$GeneSet==i)
  geneset_ind<-c(geneset_sub,geneset_ind)
}


p<-pheatmap(heatmap_geneset[geneset_ind,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)

ggsave("Heatmap_Genesets_Of_Interest_humanInMouse.pdf",plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)
```


















```{r}

label_ge1 <- unique(top_feat$all)


violin_df <- cbind(pData(geneSet_data) %>% 
                    dplyr::select(eval(contrast_factor)),
                   t(geneSet_data@assayData$exprs[label_ge1,]))



# Select expression of getTopFeatures
vexp_getTopFeatures <- exprs(geneSet_data)[label_ge1, ]

# Create expression vector for violin plot
violin_df <- cbind(pData(geneSet_data[, ]) %>% dplyr::select(eval(contrast_factor)),
                  t(vexp_getTopFeatures))

violin_df <- violin_df %>% tidyr::pivot_longer(cols=-1, names_to = "Feature", values_to = "Enrichment")

colnames(violin_df)[1] <- "contrast_factor"
violin_df$Feature <- trimNames(violin_df$Feature, 25)

# Format summary stats
violin_p_df <- filter(geneSet_results_contrast1, Feature %in% label_ge1)
violin_p_df <- violin_p_df %>% tidyr::separate(col = Comparison, into=c("group1", "group2"), sep=" vs ")
violin_p_df$FDR <- signif(violin_p_df$FDR, 3)
violin_p_df$Feature <- trimNames(violin_p_df$Feature, 25)
violin_exp_max <- ddply(violin_df, .(Feature), summarize,
                        y.position=max(Enrichment)+0.5) # +1 for safe log2
violin_p_df <- base::merge(violin_p_df, violin_exp_max, by="Feature")

p <- ggplot(violin_df, 
            aes(x=contrast_factor, y=Enrichment, fill=contrast_factor)) + 
  geom_violin(alpha=0.2) +
  geom_jitter(width=0.25, height=0, size = 1.5) + 
 scale_fill_manual(values = c("blue", "red", "grey"))  +
  facet_wrap(~Feature, scales = "free_y") +
  labs(x = eval(contrast_factor), y = "Enrichment (z-score)") +
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.2)),
                     labels = scales::number_format(accuracy = 0.01)) +
  theme_bw(base_size = 10) + 
  theme(legend.position = "bottom") +
  guides(fill=guide_legend(title = eval(contrast_factor)))

p <- p + ggprism::add_pvalue(
  violin_p_df,
  label="FDR = {FDR}", label.size = 3.6,
  y.position = violin_p_df$y.position
)


p



ggsave(p, filename = file.path(gsea_plot_dir, "gsea_violins.svg"), width=20, height=12)
```







Heatmaps of Geneset

```{r}





geneset_sub<-which(geneSet_data@featureData@data$GeneSet=="Retrograde endocannabinoid signaling")
  
gene_ids<-geneSet_data@featureData@data$TargetIds[geneset_sub]

#make gene ids seperate in vector

gene_ids<-unlist(strsplit(gene_ids,";"))

#mathc gene ids to rows in data
ind<-rownames(target_data@assayData$log_q) %in% gene_ids
  
de_heat_dat_noSub<-target_data[ind,]

annots<-de_heat_dat_noSub@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 


max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)

#color_pal<-colorRampPalette(c("black", "purple2", "yellow2"))(121)

color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(121))


#color_pal<-(colorRampPalette(c( "white","steelblue1", "firebrick1"))(121))


#color_pal<-(colorRampPalette(c("blue", "red"))(301))

heat_dat_maxmin<-de_heat_dat_noSub@assayData$log_q
  
# heat_dat_maxmin[which(heat_dat_maxmin>20)]<-20

 #heat_dat_maxmin[which(heat_dat_maxmin<1)]<-1

#color_pal<-viridis(350)


p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=3
)










ggsave("Retrograde endocannabinoid signaling.pdf",plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)


```





Automate Genset gene expression plotting


```{r}


for(i in geneset_ind){
  set_name<-geneSet_data@featureData@data$GeneSet[i]
 gene_ids<-geneSet_data@featureData@data$TargetIds[i] 
 gene_ids<-unlist(strsplit(gene_ids,";"))
 ind<-rownames(target_data@assayData$log_q) %in% gene_ids
 de_heat_dat_noSub<-target_data[ind,]
 annots<-de_heat_dat_noSub@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 

max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)
color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(121))
heat_dat_maxmin<-de_heat_dat_noSub@assayData$log_q
p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=3
)
ggsave(filename= print(paste0("Heatmap_",set_name,".pdf")),plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)

}
```





Check overlap of top DE genesets between mouse and human

```{r}

top_feat<- getTopFeatures(geneSet_results_contrast1, n_features = 50,
                            est_thr = 0.5, fdr_thr = 0.001)
saveRDS(top_feat,"top_genesets_mouse")


top_feat_human<-readRDS("top_genesets_human")


up_mouse<-unlist(top_feat[1])

up_human<-unlist(top_feat_human[1])


conserved_up<-up_human %in% up_mouse

conserved_up_list<-up_human[conserved_up]



down_mouse<-unlist(top_feat[2])

down_human<-unlist(top_feat_human[2])


conserved_down<-down_human %in% down_mouse

conserved_down_list<-down_human[conserved_down]



#now check if any don't match directionalility  between mouse and human
all_mouse<-unlist(top_feat[3])

all_human<-unlist(top_feat_human[3])

conserved_all<-all_human %in% all_mouse

conserved_all_list<-all_human[conserved_all]


conserved_overlap_up<-conserved_all_list %in% conserved_up_list 

conserved_all_list<-conserved_all_list[!conserved_overlap_up]

conserved_overlap_down<- conserved_all_list %in% conserved_down_list 

conserved_all_list<-conserved_all_list[!conserved_overlap_down]


```


Get conserved up genesets

```{r}



for(i in conserved_up_list){
  ind<-which(geneSet_data@featureData@data$GeneSet== i)
  set_name<-geneSet_data@featureData@data$GeneSet[ind]
  #use if statement if any genesets won't save due to name
  if(set_name=="Hairy, Hairy/E(SPL)"){
    set_name<-"Hairy_E_SPL"
  }
 gene_ids<-geneSet_data@featureData@data$TargetIds[ind] 
 gene_ids<-unlist(strsplit(gene_ids,";"))
 ind<-rownames(target_data@assayData$log_q) %in% gene_ids
 de_heat_dat_noSub<-target_data[ind,]
 annots<-de_heat_dat_noSub@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 

max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)
color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(121))
heat_dat_maxmin<-de_heat_dat_noSub@assayData$log_q
p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=3
)
ggsave(filename= print(paste0("Conserved_MuUp_Heatmap_",set_name,".pdf")),plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)

}






```








Get conserved down genesets


```{r}
for(i in conserved_down_list){
  ind<-which(geneSet_data@featureData@data$GeneSet== i)
  set_name<-geneSet_data@featureData@data$GeneSet[ind]
  #use if statement if any genesets won't save due to name
  if(set_name=="non-specific serine/threonine protein kinase"){
    set_name<-"non_specific_serine_threonine_protein_kinase"
  }

 gene_ids<-geneSet_data@featureData@data$TargetIds[ind] 
 gene_ids<-unlist(strsplit(gene_ids,";"))
 ind<-rownames(target_data@assayData$log_q) %in% gene_ids
 de_heat_dat_noSub<-target_data[ind,]
 annots<-de_heat_dat_noSub@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  Layer= c("5"="greenyellow","6"="mediumorchid1"),
  region= c(ACA= "lightcyan2" ,MOp="sandybrown" , MOs="maroon3")
) 

max<-max(target_data@assayData$log_q)
min<-min(target_data@assayData$log_q)
color_pal<-(colorRampPalette(c("#0092b5", "white", "#a6ce39"))(121))
heat_dat_maxmin<-de_heat_dat_noSub@assayData$log_q
p<-pheatmap(heat_dat_maxmin[,], 
scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=3
)
ggsave(filename= print(paste0("Conserved_MuDown_Heatmap_",set_name,".pdf")),plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)

}


```











Now plot just conserved genesets not individual genes
```{r}
#now check if any don't match directionalility  between mouse and human
all_mouse<-unlist(top_feat[3])

all_human<-unlist(top_feat_human[3])

conserved_all<-all_human %in% all_mouse

conserved_all_list<-all_human[conserved_all]


geneset_ind<-c()
for (i in conserved_all_list){
  geneset_sub<-which(geneSet_data@featureData@data$GeneSet==i)
  geneset_ind<-c(geneset_sub,geneset_ind)
}


annots<-geneSet_data@phenoData@data

annots<-annots[,c(3,7,22)]

ann_colors<- list(
  segment= c(NeuN="steelblue1", pSyn="firebrick1"),
  layer= c("5"="greenyellow","6"="mediumorchid1","2/3"="sandybrown")
) 

heatmap_geneset<-geneSet_data@assayData$ssgsea


label_ge1 <- unique(top_feat$all)

heatmap_geneset<-geneSet_data@assayData$ssgsea




```


```{r}




p<-pheatmap(heatmap_geneset[geneset_ind,],
            scale = "row", 
        show_rownames = TRUE, show_colnames = FALSE,
          border_color = NA,
          clustering_method = "average",
          cluster_rows = TRUE,
         cluster_cols = TRUE,
          clustering_distance_cols = "correlation",
   annotation_col = annots,
annotation_colors = ann_colors,
 #breaks = seq(min, 6 , 0.2) ,
         color = color_pal,
fontsize_row=5,
fontsize=5,
treeheight_row=10,
treeheight_col=25
)

ggsave("Heatmap_Genesets_Of_Interest_Conserved_Mu.pdf",plot=p,  width = 14, height = 14, units = "cm",path = gsea_plot_dir)
```



















PCA


```{r}

ind_cpu<- which(pData(target_data)$Region=="CPu")

#makle dataset with each AOI as rownames, each gene,and pdata as columns
PCA_data<-t(assayDataElement(object = target_data, elt = "log_q"))
PCA_data<- cbind.data.frame(pData(target_data)$area ,pData(target_data)$layer, pData(target_data)$AOINucleiCount, pData(target_data)$slide,  pData(target_data)$segment,  pData(target_data)$AllenscData, PCA_data )


#create vector of names of the pData you want in PCA. Order them quantitative, then qualitative
pca_colnames<-c("area" ,"layer", "AOINucleiCount" , "slide", "segment", "AllenscData")



#place column names
colnames(PCA_data)[1:length(pca_colnames)]<-pca_colnames


#label order
quantitative_factors<-c(1,3)

qualitative_factors<-c(2,4:6)








target_PCA<-FactoMineR::PCA(X= PCA_data, 
                ncp=10,                #number of principle components to keep in dataset
                scale.unit = TRUE,   #scales based on z-score... important for PCA, leave true
                ind.sup= NULL,
                quanti.sup= quantitative_factors,    #vector of the indexes of pheno data that is quantitative
                quali.sup = qualitative_factors ,     #vector of indexes of the pheno data that is qualitative
                row.w= NULL,         # weights for rows
                col.w= NULL,         # weights for columns
                graph=FALSE,         #whether graph should be auto displayed
                axes= c(1,2)         # which components to display
                  )
  
  

 the_prefix="PCA"

PCA_data<-get_PCA(object=target_data, elt = "log_q")


```








```{r}

PCA_plot_dir <-file.path(outdir, "PCA", "Plots")
PCA_data_dir <- file.path(outdir, "PCA", "Data")
dir.create(PCA_plot_dir,recursive = TRUE)
dir.create(PCA_data_dir, recursive = TRUE) 


p<-ggcorrplot(target_PCA$quanti.sup$cor, method="circle")+
  theme(plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())

 saveRDS(p, file=file.path(PCA_data_dir, "PCA_quanti.RDS"))
 ggsave("PCA_quanti.pdf",plot=p,  width = 30, height = 20, units = "cm",path = PCA_plot_dir)


   
 
 
 pdf(file=file.path(PCA_plot_dir, "PCA_quali.pdf"), width=5, height=4)
 
corrplot(target_PCA$quali.sup$coord, is.corr = FALSE)

dev.off()






 

 
corrplot(target_PCA$ind$coord[1:30,], is.corr = FALSE, tl.cex = 0.7)
corrplot(target_PCA$ind$coord[31:60,], is.corr = FALSE, tl.cex = 0.7)
corrplot(target_PCA$ind$coord[61:90,], is.corr = FALSE, tl.cex = 0.7)
corrplot(target_PCA$ind$coord[91:118,], is.corr = FALSE, tl.cex = 0.7)






pdf(file=file.path(PCA_plot_dir, "PCA_RNA.pdf"), width=5, height=4)
 
     pheatmap(target_PCA$var$cor)
     
dev.off()

saveRDS(p, file=file.path(PCA_data_dir, "PCA_rna.RDS"))
 ggsave("PCA_rna.pdf",plot=replayPlot(p),  width = 30, height = 20, units = "cm",path = PCA_plot_dir)
```





















```{r}

  
  colnames(target_PCA$var$coord) <- gsub("Dim.",
                                  paste0(the_prefix, "var_coords_"), colnames(target_PCA$ind$coord)) 
 # assayDataElement(object = target_data, elt ="PCA_Corr") <-target_PCA$var$coord
  
  
eigenvalues<-target_PCA$eig
barplot(eigenvalues[, 2], names.arg=(1:nrow(eigenvalues)), 
       main = "Variances",
       xlab = "Principal Components",
       ylab = "Percentage of variances",
       col ="steelblue")




dim_reduction_color_by <- pca_colnames[4] # The color factor for dim reduction; can be NULL
dim_reduction_shape_by <- pca_colnames[5] # The shape factor for dim reduction; can be NULL




ind_pca<-target_PCA[["ind"]][["coord"]]


p <- ggplot(data=target_data@phenoData@data, 
            aes(x=ind_pca[,1], y=ind_pca[,2])) +
            geom_point(aes(color=target_data@phenoData@data$segment,
                        shape=target_data@phenoData@data$Region), alpha=0.5, size=0.2) + 
            labs(x=paste0("PCA 1 (", round(target_PCA$eig[1,2]), "%)"), 
            y=paste0("PCA 2 (", round(target_PCA$eig[2,2]), "%)"),
                        title="PCA") +
            theme_bw(base_size=2) +
            theme(legend.position = "right",  
                  legend.text = element_text( size = 1.5),
                  plot.background = element_blank(),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank(),
                  )+
  scale_color_manual(values = c("NeuN"="dodgerblue2",
                               "pSyn"= "firebrick1"))
 
  p
   
  
 
  
  saveRDS(p, file=file.path(PCA_data_dir, "PCA.RDS"))
 ggsave("PCA.pdf", p,  width = 5, height = 3, units = "cm",path = PCA_plot_dir)
ggsave("PCA.png", p,  width = 10, height = 15, units = "cm",path = PCA_plot_dir)





#Now we do the alternative plots feel free to fiddle with this to plot what you want


p <- ggplot(data=target_data@phenoData@data, 
            aes(x=ind_pca[,1], y=ind_pca[,2])) +
            geom_point(aes(color=target_data@phenoData@data$Layer,
                        shape=region), alpha=0.5, size=0.2) + 
            labs(x=paste0("PCA 1 (", round(target_PCA$eig[1,2]), "%)"), 
            y=paste0("PCA 2 (", round(target_PCA$eig[2,2]), "%)"),
                        title="PCA") +
            theme_bw(base_size=2) +
            theme(legend.position = "right",  
                  legend.text = element_text( size = 1.5),
                  plot.background = element_blank(),
                  panel.grid.major = element_blank(),
                  panel.grid.minor = element_blank(),
                  )+
  scale_color_manual(values = c("5"="green",
                               "6"= "mediumorchid1"))
 
  p

ggsave("PCA_Layer.pdf", p,  width = 5, height = 3, units = "cm",path = PCA_plot_dir)



```




































Lets look at overlap with human data

```{r}


results4_FDR_human<-readRDS("results4_FDR.rds")



ind<-which(results$Color=="FDR < 0.01")

results_FDR_Mouse<- results[ind,]


hugenes<-tolower(results4_FDR_human$Gene)
hugenes<- str_to_title(hugenes)

mugenes<-results_FDR_Mouse$Gene
ind<-match(hugenes, mugenes)

match_results<-results_FDR_Mouse[ind,]


```









Here we get descriptions of the top differentially expressed genes




Sort the data based on log fold change, then extract the most differentially expressed genes
```{r}
#sort data decreasing
results_ord<-results[order(results$Estimate, decreasing = TRUE), ]

#get most pSYN enriched
Syn_Diff<-results_ord[1:200, ]

#get most NeuN enriched

#run this to get the dimensions, you'll need to subtract 200 from the row count then take those to the total row count
dim(results)


rowcount<-nrow(results_ord)

rowcount_minus<-rowcount-199


#get the genes
NeuN_diff<- results_ord[rowcount_minus:rowcount,]



```


Get description of top differentially expressed genes

```{r}
library(rentrez)

#get list of databases
entrez_dbs()

#We want to search the gene database, lets look at the searchable terms for it, anything listed here is what you can search by (e.g. you can use the gene name for gene)

entrez_db_searchable("gene")

#Lets search for our gene of interest

search_results<- c()
for (i in NeuN_diff$Gene){
  Sys.sleep(0.1)  #rate limiting so NCBI dosnt get mad
  term<- paste(i,"[GENE] AND Mus musculus[ORGN]")
  search<- entrez_search(db="gene", term= term) 
  id<-search$ids
  id<-id[1]
  a<-i
  search_id<- c(a,id)
  search_results<- rbind(search_results,search_id)
}

colnames(search_results)<-c("Gene", "ID")


multi_sums<-entrez_fetch(db="gene", id=search_results[,2], rettype = "xml")

tax_list <- XML::xmlToList(multi_sums)

diff_gene_summary<-cbind(tax_list[["Entrezgene"]][["Entrezgene_gene"]][["Gene-ref"]][["Gene-ref_locus"]], tax_list[["Entrezgene"]][["Entrezgene_summary"]])

diff_gene_summary<-c()
for (i in 1:length(tax_list)){
  A<-tax_list[[i]][["Entrezgene_gene"]][["Gene-ref"]][["Gene-ref_locus"]]
  B<- tax_list[[i]][["Entrezgene_summary"]]
 sums<- c(A,B) 
 diff_gene_summary<-rbind(diff_gene_summary,sums)
}
colnames(diff_gene_summary)<- c("gene", "summary")
rownames(diff_gene_summary)<- diff_gene_summary[,1]
write.csv(diff_gene_summary, "diff_genes_NeuN.csv")

```


Now we make it a function for ease of use
```{r}
get_gene_summaries<-function(diff_genes_df, column_num_for_gene_names, name_of_file){
search_results<- c()
blah<-diff_genes_df[,column_num_for_gene_names]
for (i in blah){
  Sys.sleep(0.1)  #rate limiting so NCBI dosnt get mad
  term<- paste(i,"[GENE] AND Mus musculus[ORGN]")
  search<- entrez_search(db="gene", term= term) 
  id<-search$ids
  a<-i
  search_id<- c(a,id)
  search_results<- rbind(search_results,search_id)
}

colnames(search_results)<-c("Gene", "ID")


multi_sums<-entrez_fetch(db="gene", id=search_results[,2], rettype = "xml")

tax_list <- XML::xmlToList(multi_sums)

diff_gene_summary<-cbind(tax_list[["Entrezgene"]][["Entrezgene_gene"]][["Gene-ref"]][["Gene-ref_locus"]], tax_list[["Entrezgene"]][["Entrezgene_summary"]])

diff_gene_summary<-c()
for (i in 1:length(tax_list)){
  A<-tax_list[[i]][["Entrezgene_gene"]][["Gene-ref"]][["Gene-ref_locus"]]
  B<- tax_list[[i]][["Entrezgene_summary"]]
 sums<- c(A,B) 
 diff_gene_summary<-rbind(diff_gene_summary,sums)
}
colnames(diff_gene_summary)<- c("gene", "summary")
rownames(diff_gene_summary)<- diff_gene_summary[,1]
write.csv(diff_gene_summary, name_of_file)
}

#example of function
get_gene_summaries(Syn_Diff , 1 , "diff_genes_pSyn.csv")
```


Import in diff genes
```{r}

diff_genes_NeuN <- read_csv("diff_genes_NeuN.csv")

diff_genes_pSyn <- read_csv("diff_genes_pSyn.csv")

```

Match genes to data and tbhen write to disk
```{r}

genes_NeuN<-full_join(diff_genes_NeuN,NeuN_diff, by= c("gene"="Gene"))

genes_pSyn<-full_join(diff_genes_pSyn,Syn_Diff, by= c("gene"="Gene"))



write.csv(genes_NeuN,"diff_genes_NeuN_foldchange.csv")

write.csv(genes_pSyn,"diff_genes_pSyn_foldchange.csv")

```








These let us know that these genes are robustly expresed in our data not needed for now

#determine expression level of differentially expressed genes
#PSYN first
```{r}
syn_genes<-Syn_Diff$Gene
orig_genes<-target_demoData@featureData@data$TargetName
targ<-   assayDataElement(object = target_demoData, elt = "log_q")
for (i in 1:length(syn_genes)){
  thing1<- syn_genes[i]
  for (ii in 1:length(orig_genes)){
    thing2<- orig_genes[ii]
    if(thing1==thing2){
      thing3<-mean(targ[ii,])
      if(thing3 < 0.5){
        vec1 <- thing2
        low_diff_genes<- c(low_diff_genes, vec1)
        cat( ii,"is low :",thing3 )
        
      }
    }
    #Uncomment below to check its really running
    #else{
      #cat(thing2," good ")
      #}
    }
    }


```



#Now for NeuN
```{r}
NeuN_genes<-NeuN_diff$Gene

for (i in 1:length(NeuN_genes)){
  thing1<- NeuN_genes[i]
  for (ii in 1:length(orig_genes)){
    thing2<- orig_genes[ii]
    if(thing1==thing2){
      thing3<-mean(targ[ii,])
      if(thing3 < 0.5){
        vec1 <- thing2
        low_diff_genes<- c(low_diff_genes, vec1)
        cat( ii,"is low :",thing3 )
        
      }
    }
    #Uncomment below to check its really running
    #else{
      #cat(thing2," good ")
      }
    }
    }


```























